{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "95d86f75-fdcd-40c8-a5a7-93bcf717702e",
   "metadata": {},
   "source": [
    "# Jupyter Notebook for Rosalinda INI Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fbe80a-f2ae-45e9-8405-78d2dc5c2e7b",
   "metadata": {},
   "source": [
    "## INI 1 - Installing Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b20cf8a-7bc5-4b9d-a143-8ffd9e97ba6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Zen of Python, by Tim Peters\n",
      "\n",
      "Beautiful is better than ugly.\n",
      "Explicit is better than implicit.\n",
      "Simple is better than complex.\n",
      "Complex is better than complicated.\n",
      "Flat is better than nested.\n",
      "Sparse is better than dense.\n",
      "Readability counts.\n",
      "Special cases aren't special enough to break the rules.\n",
      "Although practicality beats purity.\n",
      "Errors should never pass silently.\n",
      "Unless explicitly silenced.\n",
      "In the face of ambiguity, refuse the temptation to guess.\n",
      "There should be one-- and preferably only one --obvious way to do it.\n",
      "Although that way may not be obvious at first unless you're Dutch.\n",
      "Now is better than never.\n",
      "Although never is often better than *right* now.\n",
      "If the implementation is hard to explain, it's a bad idea.\n",
      "If the implementation is easy to explain, it may be a good idea.\n",
      "Namespaces are one honking great idea -- let's do more of those!\n"
     ]
    }
   ],
   "source": [
    "# display the Zen of Python\n",
    "import this"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73391690-d4cc-4ebe-b60c-365011a54038",
   "metadata": {},
   "source": [
    "## INI 2 - Variables and Some Arithmetic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12f95e7b-e3fc-4867-af0e-738b5a006ad8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "size of \"a\" side:  3\n",
      "size of \"b\" side:  3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "square on the hypotenuse: 18.0\n"
     ]
    }
   ],
   "source": [
    "# calculation of the size of a square on the hypotenuse (right-angled triangle)\n",
    "\n",
    "# input side lengths as variables\n",
    "a = int(input('size of \"a\" side: ')) # size of 'a' side\n",
    "b = int(input('size of \"b\" side: ')) # size of 'b' side\n",
    "\n",
    "# calcualate (using exponentiation)\n",
    "square = (a**2 + b**2)\n",
    "\n",
    "# print the results (of calculation)\n",
    "print('square on the hypotenuse:', square)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3315bfd7-8abf-411c-b5b6-e3cdf3d52780",
   "metadata": {},
   "source": [
    "## INI 3 - Strings and Lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63002c2d-997e-41c5-ad04-1cec318cdd34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "provide a string:  HumptyDumptysatonawallHumptyDumptyhadagreatfallAlltheKingshorsesandalltheKingsmenCouldntputHumptyDumptyinhisplaceagain.\n",
      "provide an index \"a\":  22\n",
      "provide an index \"b\":  27\n",
      "provide an index \"c\":  97\n",
      "provide an index \"d\":  102\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Humpty Dumpty\n"
     ]
    }
   ],
   "source": [
    "# slicing of a provided string based on determined positions to be sliced, not excluding, but comprising \"b\" and \"d\" indices.\n",
    "\n",
    "# input string (to be sliced)\n",
    "s = str(input('provide a string: '))\n",
    "\n",
    "# input indices (to determine the extent of slicing) \n",
    "a = int(input('provide an index \"a\": '))\n",
    "b = int(input('provide an index \"b\": '))\n",
    "c = int(input('provide an index \"c\": '))\n",
    "d = int(input('provide an index \"d\": '))\n",
    "\n",
    "# print results (of slicing)\n",
    "print(s[a:b+1], s[c:d+1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b877fd4-9698-44d2-841f-b5213a0f3107",
   "metadata": {},
   "source": [
    "## INI 4 - Conditions and Loops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "912e84f8-9fff-4633-b4b7-042d72a84cb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert integer value for \"a\":  4939\n",
      "insert integer value for \"b\":  9814\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17982688\n"
     ]
    }
   ],
   "source": [
    "# return the sum of all odd integers from a through b, inclusively.\n",
    "\n",
    "##VARIABLES\n",
    "# input integers (according to the following scheme: a < b < 10000)\n",
    "a = int(input('insert integer value for \"a\": '))\n",
    "b = int(input('insert integer value for \"b\": '))\n",
    "\n",
    "# create a list of odd numbers\n",
    "odd_list = [1, 3, 5, 7, 9]\n",
    "\n",
    "# create a variable with value 0\n",
    "odd_sum = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for loop (from a through b, inclusively)\n",
    "for num in range(a, b + 1):\n",
    "    #print(num)\n",
    "    num_list = list(str(num)) # transform a number into a list\n",
    "    #print(num_list)\n",
    "    last_num = int(num_list[len(num_list) - 1]) # get the last number\n",
    "    #print(last_num)\n",
    "\n",
    "    # if condition (if the last number is odd, add the number to the sum of previous odd numbers)\n",
    "    if last_num in odd_list: #alternatively: a % 2 == 1 <-divided by 2 and if odd, remainder is 1, applies to 1 as well\n",
    "    #if last_num % 2 == 1:\n",
    "        #print('odd')\n",
    "        odd_sum = odd_sum + num\n",
    "    #else:\n",
    "        #print('even')\n",
    "        \n",
    "# print results (of conditional calculation)\n",
    "print(odd_sum)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767e0d04-544b-47cb-b695-fb836a9ce150",
   "metadata": {},
   "source": [
    "## INI 5 - Working with Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "557be151-ec4d-4001-aeb2-f3cd1f823f0a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "53bf4e1a-7f1c-4cb1-bb1a-c50463a82f10",
   "metadata": {},
   "source": [
    "## INI 6 - Dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "cfc1a062-00a1-421e-a4d5-c96c5cb01f94",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We 1\n",
      "tried 3\n",
      "list 1\n",
      "and 1\n",
      "we 2\n",
      "dicts 1\n",
      "also 1\n",
      "Zen 1\n"
     ]
    }
   ],
   "source": [
    "# return the number of occurrences of each word in variable 's' (string), where words are separated by spaces; words are case-sensitive; lines in the output can be in any order\n",
    "\n",
    "## VARIABLES\n",
    "# input string and dictionary-based accumulator variable\n",
    "s = 'We tried list and we tried dicts also we tried Zen'\n",
    "#s = str(input('insert string:'))\n",
    "word_count = {}\n",
    "\n",
    "## PROGRAMME - extracting information\n",
    "# for loop (from the first word from the list to the last one)\n",
    "for word in s.split(): # hint: use (' ') for space or other delimiter character\n",
    "    #if word exists in the dictionary, then add 1 to its value\n",
    "    if word in word_count:\n",
    "        word_count[word] = word_count[word] + 1\n",
    "        # alternatively\n",
    "        #word_count.update({word : word_count[word] + 1})\n",
    "    #if not, then create new key-value pair with a value 1\n",
    "    else:\n",
    "        word_count[word] = 1\n",
    "        # alternatively\n",
    "        #word_count.update({word : 1})\n",
    "\n",
    "#print(word_count) # debugging\n",
    "#word_count = dict(sorted(word_count.items())) # alphabetically ordered dictionary\n",
    "\n",
    "## VARIABLES\n",
    "# integer-based accumulator variable\n",
    "index = 0\n",
    "\n",
    "## PROGRAMME - processing information\n",
    "# for loop (from the first key-value pair from the dictionary to the last one)\n",
    "for word in word_count: \n",
    "    # print words and their occurence number\n",
    "    print(list(word_count)[index], word_count[word])\n",
    "    # add one to the accumulator variable\n",
    "    index = index + 1\n",
    "\n",
    "# hint: alternative easy way\n",
    "# for word, value in word_count.items():\n",
    "#     print(word, value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b353b205-3799-445a-b19b-5d3acef2d702",
   "metadata": {},
   "source": [
    "## INI DNA - Counting Nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "db7b261c-a701-49c1-84c7-e0793210fcb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "DNA string: GACTTGGCGATCCAATTGACTCTCGCTACAGGAAATTAGCGACGCGGGTCGTCGCCACAAAAGGCCCCCATGCACTACTAGATGTCCCTTGTTCTCCTCTGAATCACATACCGGAACGCCATACATAACCTGGAAGCCGTTTTCGTAATCTGGAAATTAGACAAAACGCGGAGAGCGTTGCTAACTCTCCCGGGGTCGTCCTGTAGACAAGTTGCACAAAGCCTGCATAATCCCTCATCTTTATACGAATATATGGAGTTACGGCGAATAGTAAGGGTTCTAGTCTAGGATAGGATAGGCCCAAGGCAAGTGGTCAGTCTGGGCCTCCTGGTGAGGTAGCAAGGGGCGTGTTTATATCATAGTCCGAATAGTGGTAACGTCGTCAGCGGCGATGCATGCAAACGAGTCCCGCTGTAAGTCGCCTCGATCTTCTTTATGCTCCTAGCCTGACAAGGTACTGTATATAATGCTCCACGACATGCCGAAATGCTCGCTAAGCACGGCCGTAATACGAGAACTTCCCTATGAACAGTTTATGACGACGGAGGGTCCACGATCCACCACTAAAGTCAAGTATGCCAGTCGTGGGATAACTTCACCGATTCAAGATTCGAGTGAGACACAAGTACACGTAAAATGGGAGATTCAACCTCTTGTCAGCACAAGGATGCGCTTAATTTGTACACTAAAGATGGCCCTAGCCGCAGCCCGAAAACGCTTACTGAGCTCGATGTAGCGTATGCAGAGGTATCACGTTAATGGCGTCACGTGGGTCTCGCAAGTAGTGAGGTTCAAGTGTTAATTACTAACGTCTCGCGCCGATTATAATGGCATTTGATGCGCTGTATATAGTAGTGGTTAGATAAATACACCTAGATTCCCAGTGAAGCCTCGTACTATCTTTTATTAATCTAGCCTTTTCCATACGAGGAGCTACTATCACACTAGCAGAATACAACGTAATCGTGCTTTCCTATAATGCCAAGACACTCTTT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "272 239 232 252\n"
     ]
    }
   ],
   "source": [
    "# insert the DNA sequnce (as a string)\n",
    "#s = str(input('insert the DNA sequence: '))\n",
    "\n",
    "# make a (DNA) list with counts per each nucleotide base (\"A\", \"C\", \"G\", \"T\")\n",
    "#DNA = [s.count('A'), s.count('C'), s.count('G'), s.count('T')]\n",
    "\n",
    "# print the counts (for nucleotide base - \"A\", \"C\", \"G\", \"T\")\n",
    "\n",
    "\n",
    "## print('A occurs: ', DNA[0], 'times',\n",
    "##       'C occurs: ', DNA[1], 'times',\n",
    "##       'G occurs: ', DNA[2], 'times',\n",
    "##       'T occurs: ', DNA[3], 'times',)\n",
    "\n",
    "#print(int(DNA[0]), int(DNA[1]), int(DNA[2]), int(DNA[3]))\n",
    "\n",
    "\n",
    "\n",
    "##BONUS - Rosalind DNA, but use a dictionary format\n",
    "\n",
    "# input\n",
    "#AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC\n",
    "#s = str(input('DNA string:'))\n",
    "# s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'\n",
    "\n",
    "# s_dic = {'A': s.count('A'),\n",
    "#          'C': s.count('C'),\n",
    "#          'G': s.count('G'),\n",
    "#          'T': s.count('T')\n",
    "#         }\n",
    "         \n",
    "# print(s_dic)\n",
    "\n",
    "\n",
    "\n",
    "##BONUS - Rosalind DNA, but use a loop\n",
    "\n",
    "##VARIABLES\n",
    "#s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'\n",
    "s = str(input('DNA string:'))\n",
    "count = {}\n",
    "##PROGRAMME\n",
    "for nucleotide_base in list(s):\n",
    "    if nucleotide_base in count:\n",
    "        count[nucleotide_base] = count[nucleotide_base] + 1\n",
    "        #count.update({nucleotide_base : count[nucleotide_base] + 1})\n",
    "    else:\n",
    "        count[nucleotide_base] = 1\n",
    "        #count.update({nucleotide_base : 1})\n",
    "\n",
    "count = dict(sorted(count.items())) # alphabetical sorting\n",
    "\n",
    "print(list(count.values())[0], list(count.values())[1], list(count.values())[2], list(count.values())[3]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61328818-8801-472b-872b-d01ae539064c",
   "metadata": {},
   "source": [
    "## INI RNA - Transcribing DNA into RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f7a405f-5ea6-496d-81a1-ec6de82a87ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert the DNA sequence:  CTACCATCTTAAAGGTACCGGGCAATCGCCTACTGAGGCCTGGTTAGTAATTGGGTTCCCACAAATAGCTCTATAACCCGTGCCTATATCCCGCTATTAATTCAAGCGTGCGTTGATCACGAAGCGGTGAATCTTCAGCACGGGAAGCCTTTGAAGCGAGGACTAGGAAAGCTTTTTGTTGTCAGGCGGTCATACCCGTCTAACTTGACAAAGACAGCCTAGAATCTAGCGTCTCACCATGGTGAGATGCAATTAGATATCTTAAAATTTCTCGGGCCGTCGAACATGGGGCCCATGACTAAATTCCATCTGCTGCACCCAACGATGAACATCTTTCGTAGGTTGCCTCCGAACAAAGTTAGTGCTGGACTAAGAACTGTAACTCCCGTGCGGACGCTCGCCAGTAGTACGCATTGTTGGCGTGCCTCATCTGTACCCCCCCTGACATTTGCGCCGTGTTTCTGGTCGCTGATTGAGATAGGCCTGGGATTTGTTGTACCCCCGCCTTTAGGAGTTCGGTCCCCCGCAAGAGAACCGTGGGTCCGAAAACGGCAGGTTCCCCAAGAAGCGGCCCGCCAATAGCGTACTAGCTTACGTAAAATGCCGAAGGGCTCCTCGGCCAGATAGAAGCTTGCTCTTCGTGCCATTCCAGCCGGTCAGTGTTTATAGGGTCTCAAATGGCGATCAGTTTTGCCATTACGCGGGTCCCGATCCTGGAGCATTTCTCGACAGCGTAATGACACATGAGTCGCATCGTACTGGTGAATAGGTGAAGGTTCAAGGGAGGTCTTCTCTGGGGTGTGAGACAACAGGTGAAATCCGAGTTCGACCTGTCGATACAGACTCACTCAAAAAAAATATGCTACTGTCCGCAACCTTGAGATCGCGGTAGAAACACGGGTAGTGCGGTTATCCATCGCCTGGTATCTCCGGTGAGACGGGTAGCCTTGTACCATAGTTTAAGCTCCAGCAATC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CUACCAUCUUAAAGGUACCGGGCAAUCGCCUACUGAGGCCUGGUUAGUAAUUGGGUUCCCACAAAUAGCUCUAUAACCCGUGCCUAUAUCCCGCUAUUAAUUCAAGCGUGCGUUGAUCACGAAGCGGUGAAUCUUCAGCACGGGAAGCCUUUGAAGCGAGGACUAGGAAAGCUUUUUGUUGUCAGGCGGUCAUACCCGUCUAACUUGACAAAGACAGCCUAGAAUCUAGCGUCUCACCAUGGUGAGAUGCAAUUAGAUAUCUUAAAAUUUCUCGGGCCGUCGAACAUGGGGCCCAUGACUAAAUUCCAUCUGCUGCACCCAACGAUGAACAUCUUUCGUAGGUUGCCUCCGAACAAAGUUAGUGCUGGACUAAGAACUGUAACUCCCGUGCGGACGCUCGCCAGUAGUACGCAUUGUUGGCGUGCCUCAUCUGUACCCCCCCUGACAUUUGCGCCGUGUUUCUGGUCGCUGAUUGAGAUAGGCCUGGGAUUUGUUGUACCCCCGCCUUUAGGAGUUCGGUCCCCCGCAAGAGAACCGUGGGUCCGAAAACGGCAGGUUCCCCAAGAAGCGGCCCGCCAAUAGCGUACUAGCUUACGUAAAAUGCCGAAGGGCUCCUCGGCCAGAUAGAAGCUUGCUCUUCGUGCCAUUCCAGCCGGUCAGUGUUUAUAGGGUCUCAAAUGGCGAUCAGUUUUGCCAUUACGCGGGUCCCGAUCCUGGAGCAUUUCUCGACAGCGUAAUGACACAUGAGUCGCAUCGUACUGGUGAAUAGGUGAAGGUUCAAGGGAGGUCUUCUCUGGGGUGUGAGACAACAGGUGAAAUCCGAGUUCGACCUGUCGAUACAGACUCACUCAAAAAAAAUAUGCUACUGUCCGCAACCUUGAGAUCGCGGUAGAAACACGGGUAGUGCGGUUAUCCAUCGCCUGGUAUCUCCGGUGAGACGGGUAGCCUUGUACCAUAGUUUAAGCUCCAGCAAUC\n"
     ]
    }
   ],
   "source": [
    "# insert the DNA sequence (as a 't' string)\n",
    "t = str(input('insert the DNA sequence: '))\n",
    "\n",
    "\n",
    "# return the transrcibed DNA sequence (as a 'u' string, that has 'A', 'C', 'G' and 'U' nucleotide bases instead of 'A', 'C', 'G' and 'T')\n",
    "u = t.replace('T', 'U')\n",
    "\n",
    "# print the RNA sequence\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a83862bb-072f-49d2-bb4d-a69cb458c0dd",
   "metadata": {},
   "source": [
    "## HAMM - Counting Point Mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a6a8c0-c6fd-4f05-9e3b-28dc71067947",
   "metadata": {},
   "source": [
    "### Theory\n",
    "\n",
    "A mutation is a mistake which occurs during the creation/copying of a nucleic acid, in particular DNA. Sum of micro-mutations over many generations might lead to positive macro effects.\n",
    "\n",
    "The simpliest and most common type of nucleic acid mutation is a point mutation, which replaces one base with another a a single nucleotide. In the case of DNA, a point mutation must change the complementary base accordingly.\n",
    "\n",
    "Two DNA strands taken from different organism or species genomes are homologous if they share a recent ancestor; thus, counting the number of bases at which homologous strands differ provides us with the minimum number of point mutations that could have occurred on the evolutionary path between the two strands. Minimising the number of point mutations separating two species needs to be done, since the principle of parsimony demands that evolutionary histories should be be as simply explained as possible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "cfd93272-9346-44d3-b076-c950f532a3b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "# return the hamming distance (number of mismatching nucleotide bases)\n",
    "\n",
    "##VARIABLE\n",
    "s = 'GAGCCTACTAACGGGAT'\n",
    "#s = str(input('first DNA sequence:')\n",
    "t = 'CATCGTAATGACGGCCT'\n",
    "#t = str(input('second DNA sequence:')\n",
    "d = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for particular position (index) in range of length of the string, do\n",
    "for position in range(len(s)): # alternatively a while loop\n",
    "    # if nucleotide base at the particular position in 's' string is not the same as the nucleotide base at the aprticular position in 't' string, then do\n",
    "    if s[position] != t[position]:\n",
    "        # add 1 to hamming distance\n",
    "        d = d + 1 # alternatively +=\n",
    "    #if not, then continue to with another iteration\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "# print results\n",
    "print(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d338f783-477a-4528-8bba-a18b738bb869",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
