{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "95d86f75-fdcd-40c8-a5a7-93bcf717702e",
   "metadata": {},
   "source": [
    "# Jupyter Notebook for Rosalinda INI Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fbe80a-f2ae-45e9-8405-78d2dc5c2e7b",
   "metadata": {},
   "source": [
    "## INI 1 - Installing Python (Import)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b20cf8a-7bc5-4b9d-a143-8ffd9e97ba6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Zen of Python, by Tim Peters\n",
      "\n",
      "Beautiful is better than ugly.\n",
      "Explicit is better than implicit.\n",
      "Simple is better than complex.\n",
      "Complex is better than complicated.\n",
      "Flat is better than nested.\n",
      "Sparse is better than dense.\n",
      "Readability counts.\n",
      "Special cases aren't special enough to break the rules.\n",
      "Although practicality beats purity.\n",
      "Errors should never pass silently.\n",
      "Unless explicitly silenced.\n",
      "In the face of ambiguity, refuse the temptation to guess.\n",
      "There should be one-- and preferably only one --obvious way to do it.\n",
      "Although that way may not be obvious at first unless you're Dutch.\n",
      "Now is better than never.\n",
      "Although never is often better than *right* now.\n",
      "If the implementation is hard to explain, it's a bad idea.\n",
      "If the implementation is easy to explain, it may be a good idea.\n",
      "Namespaces are one honking great idea -- let's do more of those!\n"
     ]
    }
   ],
   "source": [
    "# display the Zen of Python\n",
    "import this"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73391690-d4cc-4ebe-b60c-365011a54038",
   "metadata": {},
   "source": [
    "## INI 2 - Variables and Some Arithmetic (Mathematical Operation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12f95e7b-e3fc-4867-af0e-738b5a006ad8",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "Interrupted by user",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# calculation of the size of a square on the hypotenuse (right-angled triangle)\u001b[39;00m\n\u001b[32m      2\u001b[39m \n\u001b[32m      3\u001b[39m \u001b[38;5;66;03m# input side lengths as variables\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m a = \u001b[38;5;28mint\u001b[39m(\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43msize of \u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43ma\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m side: \u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m) \u001b[38;5;66;03m# size of 'a' side\u001b[39;00m\n\u001b[32m      5\u001b[39m b = \u001b[38;5;28mint\u001b[39m(\u001b[38;5;28minput\u001b[39m(\u001b[33m'\u001b[39m\u001b[33msize of \u001b[39m\u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m side: \u001b[39m\u001b[33m'\u001b[39m)) \u001b[38;5;66;03m# size of 'b' side\u001b[39;00m\n\u001b[32m      7\u001b[39m \u001b[38;5;66;03m# calcualate (using exponentiation)\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/carpentries/lib/python3.14/site-packages/ipykernel/kernelbase.py:1396\u001b[39m, in \u001b[36mKernel.raw_input\u001b[39m\u001b[34m(self, prompt)\u001b[39m\n\u001b[32m   1394\u001b[39m     msg = \u001b[33m\"\u001b[39m\u001b[33mraw_input was called, but this frontend does not support input requests.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1395\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m StdinNotImplementedError(msg)\n\u001b[32m-> \u001b[39m\u001b[32m1396\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_input_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1397\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1398\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_get_shell_context_var\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_shell_parent_ident\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1399\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mget_parent\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mshell\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1400\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpassword\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m   1401\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/carpentries/lib/python3.14/site-packages/ipykernel/kernelbase.py:1441\u001b[39m, in \u001b[36mKernel._input_request\u001b[39m\u001b[34m(self, prompt, ident, parent, password)\u001b[39m\n\u001b[32m   1438\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[32m   1439\u001b[39m     \u001b[38;5;66;03m# re-raise KeyboardInterrupt, to truncate traceback\u001b[39;00m\n\u001b[32m   1440\u001b[39m     msg = \u001b[33m\"\u001b[39m\u001b[33mInterrupted by user\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1441\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m(msg) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m   1442\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[32m   1443\u001b[39m     \u001b[38;5;28mself\u001b[39m.log.warning(\u001b[33m\"\u001b[39m\u001b[33mInvalid Message:\u001b[39m\u001b[33m\"\u001b[39m, exc_info=\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: Interrupted by user"
     ]
    }
   ],
   "source": [
    "# calculation of the size of a square on the hypotenuse (right-angled triangle)\n",
    "\n",
    "# input side lengths as variables\n",
    "a = int(input('size of \"a\" side: ')) # size of 'a' side\n",
    "b = int(input('size of \"b\" side: ')) # size of 'b' side\n",
    "\n",
    "# calcualate (using exponentiation)\n",
    "square = (a**2 + b**2)\n",
    "\n",
    "# print the results (of calculation)\n",
    "print('square on the hypotenuse:', square)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3315bfd7-8abf-411c-b5b6-e3cdf3d52780",
   "metadata": {},
   "source": [
    "## INI 3 - Strings and Lists (Slice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63002c2d-997e-41c5-ad04-1cec318cdd34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# slicing of a provided string based on determined positions to be sliced, not excluding, but comprising \"b\" and \"d\" indices.\n",
    "\n",
    "# input string (to be sliced)\n",
    "s = str(input('provide a string: '))\n",
    "\n",
    "# input indices (to determine the extent of slicing) \n",
    "a = int(input('provide an index \"a\": '))\n",
    "b = int(input('provide an index \"b\": '))\n",
    "c = int(input('provide an index \"c\": '))\n",
    "d = int(input('provide an index \"d\": '))\n",
    "\n",
    "# print results (of slicing)\n",
    "print(s[a:b+1], s[c:d+1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b877fd4-9698-44d2-841f-b5213a0f3107",
   "metadata": {},
   "source": [
    "## INI 4 - Conditions and Loops (Loop/Con - is in?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "912e84f8-9fff-4633-b4b7-042d72a84cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return the sum of all odd integers from a through b, inclusively.\n",
    "\n",
    "##VARIABLES\n",
    "# input integers (according to the following scheme: a < b < 10000)\n",
    "a = int(input('insert integer value for \"a\": '))\n",
    "b = int(input('insert integer value for \"b\": '))\n",
    "\n",
    "# create a list of odd numbers\n",
    "odd_list = [1, 3, 5, 7, 9]\n",
    "\n",
    "# create a variable with value 0\n",
    "odd_sum = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for loop (from a through b, inclusively)\n",
    "for num in range(a, b + 1):\n",
    "    #print(num)\n",
    "    num_list = list(str(num)) # transform a number into a list\n",
    "    #print(num_list)\n",
    "    last_num = int(num_list[len(num_list) - 1]) # get the last number\n",
    "    #print(last_num)\n",
    "\n",
    "    # if condition (if the last number is odd, add the number to the sum of previous odd numbers)\n",
    "    if last_num in odd_list: #alternatively: a % 2 == 1 <-divided by 2 and if odd, remainder is 1, applies to 1 as well\n",
    "    #if last_num % 2 == 1:\n",
    "        #print('odd')\n",
    "        odd_sum = odd_sum + num\n",
    "    #else:\n",
    "        #print('even')\n",
    "        \n",
    "# print results (of conditional calculation)\n",
    "print(odd_sum)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c35ebbea-945d-4410-bf89-7b6e563f4e69",
   "metadata": {},
   "source": [
    "## INI4 - Conditions and Loops (Loop/Con - is odd/even?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b91b24-67d1-4aa4-8dc5-ec3fa335a064",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return the sum of all odd integers from a through b, inclusively.\n",
    "\n",
    "##VARIABLES\n",
    "# input integers (according to the following scheme: a < b < 10000)\n",
    "a = 100\n",
    "b = 200\n",
    "odd_sum = 0\n",
    "\n",
    "for num in range(a, b + 1):\n",
    "    if num % 2 != 0:\n",
    "        odd_sum = odd_sum + num\n",
    "\n",
    "print(odd_sum)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767e0d04-544b-47cb-b695-fb836a9ce150",
   "metadata": {},
   "source": [
    "## INI 5 - Working with Files (Slice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "557be151-ec4d-4001-aeb2-f3cd1f823f0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Some things in life are bad, they can really make you mad\n",
      "Other things just make you swear and curse\n",
      "When you're chewing on life's gristle, don't grumble give a whistle\n",
      "This will help things turn out for the best\n",
      "Always look on the bright side of life\n",
      "Always look on the right side of life\n",
      "If life seems jolly rotten, there's something you've forgotten\n",
      "And that's to laugh and smile and dance and sing\n",
      "When you're feeling in the dumps, don't be silly, chumps\n",
      "Just purse your lips and whistle, that's the thing\n",
      "So, always look on the bright side of death\n",
      "Just before you draw your terminal breath\n",
      "Life's a counterfeit and when you look at it\n",
      "Life's a laugh and death's the joke, it's true\n",
      "You see, it's all a show, keep them laughing as you go\n",
      "Just remember the last laugh is on you\n",
      "Always look on the bright side of life\n",
      "And always look on the right side of life\n",
      "Always look on the bright side of life\n",
      "And always look on the right side of life\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# return a file containing all the even-numbered lines form the original file and assume 1-based numbering of lines\n",
    "\n",
    "# input string: 'Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat'\n",
    "\n",
    "# create an example file\n",
    "# file_in = open('../../rosalind_INI4.txt', 'w')\n",
    "# file_in.write('Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat')\n",
    "# file_in.close()\n",
    "\n",
    "##VARIABLES\n",
    "# input file path\n",
    "#input_path = str(input('insert file path for input text file:'))\n",
    "input_path = '../../rosalind_INI5.txt'\n",
    "# output file path\n",
    "#output_path = str(input('insert file path for output text file:'))\n",
    "output_path = '../../rosalind_INI5_output.txt'\n",
    "\n",
    "# load text\n",
    "file_in = open(input_path, 'r')\n",
    "# create output file\n",
    "file_out = open(output_path, 'w')\n",
    "\n",
    "##PROGRAMME\n",
    "# extract the number of lines\n",
    "#line_count = len(file_in.readlines()) + 1\n",
    "# reload input file\n",
    "#file_in = open(input_path, 'r')\n",
    "# extract even-numbered lines (odd numbers in slicing, because starts from 0)\n",
    "output = file_in.readlines()[1::2] \n",
    "\n",
    "##NOTE\n",
    "# if length is defined ([1:line_count:2]), after line_count, input file needs to be reloaded\n",
    "\n",
    "# convert list of strings into a unified string\n",
    "string = ''.join(output)\n",
    "# write output into the output file\n",
    "file_out.write(string)\n",
    "# close file modifications\n",
    "file_out.close()\n",
    "\n",
    "# read output file\n",
    "file_out = open(output_path, 'r')\n",
    "# transform list into a string\n",
    "result = ''.join(file_out.readlines())\n",
    "# print results\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d275f627-3293-498e-9138-586fbf8f3dac",
   "metadata": {},
   "source": [
    "## INI 5 - Working with Files (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18ee6d31-14bf-4154-87a4-33d8cfed8958",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return a file containing all the even-numbered lines form the original file and assume 1-based numbering of lines\n",
    "\n",
    "# input string: 'Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat'\n",
    "\n",
    "# create an example file\n",
    "# file_in = open('../../rosalind_INI4.txt', 'w')\n",
    "# file_in.write('Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat')\n",
    "# file_in.close()\n",
    "\n",
    "##VARIABLES\n",
    "# input file path\n",
    "#input_path = str(input('insert file path for input text file:'))\n",
    "input_path = '../../rosalind_INI5.txt'\n",
    "# output file path\n",
    "#output_path = str(input('insert file path for output text file:'))\n",
    "output_path = '../../rosalind_INI5_output.txt'\n",
    "\n",
    "# load text\n",
    "file_in = open(input_path, 'r')\n",
    "# create output file\n",
    "file_out = open(output_path, 'w')\n",
    "# create an accumulator variable\n",
    "position = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for each line in text file\n",
    "for line in file_in:\n",
    "    # if position is odd\n",
    "    if position % 2 == 1:\n",
    "        # write the line into output file\n",
    "        file_out.write(line)\n",
    "        # add 1 to position\n",
    "        position += 1\n",
    "    # otherwise - does not need to be included, if + add 1 would be enough\n",
    "    else:\n",
    "        # add 1 to position \n",
    "        position += 1\n",
    "        # continue with another iteration\n",
    "        continue\n",
    "\n",
    "# close writing into the output file\n",
    "file_out.close()\n",
    "\n",
    "# read output file\n",
    "file_out = open(output_path, 'r')\n",
    "# transform list into a string\n",
    "result = ''.join(file_out.readlines())\n",
    "# print results\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bf4e1a-7f1c-4cb1-bb1a-c50463a82f10",
   "metadata": {},
   "source": [
    "## INI 6 - Dictionaries (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc1a062-00a1-421e-a4d5-c96c5cb01f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return the number of occurrences of each word in variable 's' (string), where words are separated by spaces; words are case-sensitive; lines in the output can be in any order\n",
    "\n",
    "## VARIABLES\n",
    "# input string and dictionary-based accumulator variable\n",
    "#s = 'We tried list and we tried dicts also we tried Zen'\n",
    "s = str(input('insert string:'))\n",
    "word_count = {}\n",
    "\n",
    "## PROGRAMME - extracting information\n",
    "# for loop (from the first word from the list to the last one)\n",
    "for word in s.split(): # hint: use (' ') for space or other delimiter character\n",
    "    #if word exists in the dictionary, then add 1 to its value\n",
    "    if word in word_count:\n",
    "        word_count[word] = word_count[word] + 1 # alternatively: +=\n",
    "        # alternatively\n",
    "        #word_count.update({word : word_count[word] + 1})\n",
    "    #if not, then create new key-value pair with a value 1\n",
    "    else:\n",
    "        word_count[word] = 1\n",
    "        # alternatively\n",
    "        #word_count.update({word : 1})\n",
    "\n",
    "#print(word_count) # debugging\n",
    "#word_count = dict(sorted(word_count.items())) # alphabetically ordered dictionary\n",
    "\n",
    "## VARIABLES\n",
    "# integer-based accumulator variable\n",
    "index = 0\n",
    "\n",
    "## PROGRAMME - processing information\n",
    "# for loop (from the first key-value pair from the dictionary to the last one)\n",
    "for word in word_count: \n",
    "    # print words and their occurence number\n",
    "    print(list(word_count)[index], word_count[word])\n",
    "    # add one to the accumulator variable\n",
    "    index = index + 1\n",
    "\n",
    "# hint: alternative easy way\n",
    "# for word, value in word_count.items():\n",
    "#     print(word, value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b353b205-3799-445a-b19b-5d3acef2d702",
   "metadata": {},
   "source": [
    "## DNA - Counting Nucleotides (Method; Loop/Con; Dictionary/Method)\n",
    "https://rosalind.info/problems/dna/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75546f1a-b5e4-46d0-9b82-1ffdddc9301e",
   "metadata": {},
   "source": [
    "### Theory\n",
    "Considering the life and living organic material, the cell is recognised as the building block of life. The centre of the activity is considered to be in the nucles (only in eukaryotic cells, where it is present). The nucleus is filled with macromolecules called chromatin. During eukaryotic cell division (mitosis), most of the chromatin condenses into long, thin strings called chromosomes.\n",
    "\n",
    "One class of the macromolecules contained in chromatin are called nucleic acids. Nucleic acids are polymers, or repeating chains of smaller, similarly structured molecules known as monomers. Because of their tendency to be long and thin, nucleic acid polymers are commonly called strands.\n",
    "\n",
    "The nucleic acid monomer is called a nucleotide and is used as a unit of strand length (abbreviated to nt). Each nucleotide is formed of three parts: a sugar molecule, a negatively charged ion called a phosphate, and a compound called a nucleobase (\"base\" for short). Polymerisation is achieved as the sugar of one nucleotide bonds to the phosphate of the next nucleotide in the chain, which forms a sugar-phosphate backbone for the nucleic acid strand. A key point is that the nucleotides of a specific type of nucleic acid always contain the same sugar and phosphate molecules, and they differ only in their choice of base. Thus, one strand of a nucleic acid can be differentiated from another based solely on the order of its bases; this ordering of bases defines a nucleic acid's primary structure.\n",
    "\n",
    "A strand of deoxyribose nucleic acid (DNA), in which the sugar is called deoxyribose, posseses only four choices for nucleobases, molecules called adenine (A), cytosine (C), guanine (G), and thymine (T).\n",
    "\n",
    "DNA is found in all living organisms on Earth, including bacteria; it is even found in many viruses (which are often considered to be nonliving). Because of its importance, we reserve the term genome to refer to the sum total of the DNA contained in an organism's chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22a533c3-340d-4d00-a52f-46544c3c9a3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db7b261c-a701-49c1-84c7-e0793210fcb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert the DNA sequnce (as a string)\n",
    "#s = str(input('insert the DNA sequence: '))\n",
    "\n",
    "# make a (DNA) list with counts per each nucleotide base (\"A\", \"C\", \"G\", \"T\")\n",
    "#DNA = [s.count('A'), s.count('C'), s.count('G'), s.count('T')]\n",
    "\n",
    "# print the counts (for nucleotide base - \"A\", \"C\", \"G\", \"T\")\n",
    "\n",
    "\n",
    "## print('A occurs: ', DNA[0], 'times',\n",
    "##       'C occurs: ', DNA[1], 'times',\n",
    "##       'G occurs: ', DNA[2], 'times',\n",
    "##       'T occurs: ', DNA[3], 'times',)\n",
    "\n",
    "#print(int(DNA[0]), int(DNA[1]), int(DNA[2]), int(DNA[3]))\n",
    "\n",
    "\n",
    "\n",
    "##BONUS - Rosalind DNA, but use a dictionary format\n",
    "\n",
    "# input\n",
    "#AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC\n",
    "#s = str(input('DNA string:'))\n",
    "# s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'\n",
    "\n",
    "# s_dic = {'A': s.count('A'),\n",
    "#          'C': s.count('C'),\n",
    "#          'G': s.count('G'),\n",
    "#          'T': s.count('T')\n",
    "#         }\n",
    "         \n",
    "# print(s_dic)\n",
    "\n",
    "\n",
    "\n",
    "##BONUS - Rosalind DNA, but use a loop\n",
    "\n",
    "##VARIABLES\n",
    "#s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'\n",
    "s = str(input('DNA string:'))\n",
    "count = {}\n",
    "##PROGRAMME\n",
    "for nucleotide_base in list(s):\n",
    "    if nucleotide_base in count:\n",
    "        count[nucleotide_base] = count[nucleotide_base] + 1\n",
    "        #count.update({nucleotide_base : count[nucleotide_base] + 1})\n",
    "    else:\n",
    "        count[nucleotide_base] = 1\n",
    "        #count.update({nucleotide_base : 1})\n",
    "\n",
    "count = dict(sorted(count.items())) # alphabetical sorting\n",
    "\n",
    "print(list(count.values())[0], list(count.values())[1], list(count.values())[2], list(count.values())[3]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61328818-8801-472b-872b-d01ae539064c",
   "metadata": {},
   "source": [
    "## RNA - Transcribing DNA into RNA (Method)\n",
    "https://rosalind.info/problems/rna/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f39fd0e-bfbe-4871-9885-15c4da96f698",
   "metadata": {},
   "source": [
    "### Theory\n",
    "RNA is a second nucleic acid existing alongside DNA in the chromatin and contains a different sugar called ribose, which came to be known as ribose nucleic acid (RNA). RNA differs from DNA in that it contains a base called uracil in place of thymine. RNA is present in the cells of all life forms on Earth.\n",
    "\n",
    "The structure of DNA and RNA is similar because DNA seves as a blueprint for the creation of special sort of RNA, called messenger RNA (mRNA). mRNA is created during RNA transcription, during which a strand of DNA is used as a template for constructing a strand of RNA by copying nucleotides one at a time, where uracil is used in place of thymine.\n",
    "\n",
    "In eukaryotes, DNA remains in the nucleus, while RNA can enter the far reaches of the cell to carry out DNA's instructions.\n",
    "\n",
    "\n",
    "An RNA string is a string that represents the order of nucleobases along a strand of RNA. For this reason, RNA strings are constructed from the alphabet {A, C, G, U}, whose symbols represent the bases adenine, cytosine, guanine, and uracil. For example, \"AAGAUGCCGU\" is an RNA string of length 10."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f7a405f-5ea6-496d-81a1-ec6de82a87ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert the DNA sequence (as a 't' string)\n",
    "t = str(input('insert the DNA sequence: '))\n",
    "\n",
    "\n",
    "# return the transrcibed DNA sequence (as a 'u' string, that has 'A', 'C', 'G' and 'U' nucleotide bases instead of 'A', 'C', 'G' and 'T')\n",
    "u = t.replace('T', 'U')\n",
    "\n",
    "# print the RNA sequence\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e1349a1-b278-44a0-bc38-eee992033e76",
   "metadata": {},
   "source": [
    "## RNA - Transcribing DNA into RNA (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a0f46008-b16e-48cd-9e57-9b760fbc2629",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert the DNA sequence:  AGCAAGTAGCAATAGTGGGCGTGAACACTGCATGAAGCTCAGGGAATGGTTCCTGGAGTTTTATTTTAGCACACACGATAAGGTTAATAGACCTCACTGTCAAGTTCGATCGTCGAGAAGTGAAGCAAAAACGGTATCCTCGAGTGGGAAGCTCGGCACTCGCCCCTTGCACGGCGGGAAACCCACGAACACTTGTTTTCCATATGCGGCACAGACTTAACCTTAGCGTGTGAGTCGCTTCAGTACAAGATCATCGAGACAGTGCACACTGGATTAAACGTCAACACATGTGCCGACACTACCTTGGCAGTCTGTATGGCGAACCATCGTTCAAAACGGGGACTAAACGGTCAAGGCAGTAGATGGTTACGAGAGTGCTAGGATCGTGGCCAATTCCCACACATACTTGCGAGTTTGGTTTTTCACACTCGTGTGTAGAATGGGATACGCAACCAAAATCTACCTGCATACGCGCCTCATCTCGATTGTCGAAATGGAGTGTTTGAAGGTACGTTTGCCAGCGACCTTTGACAGAGTTATCGCCCTAAAGTCCTCATACCACGAGCCAGACTCACACTCCAAGAATGCCTTCGTGGCGCTGCCGCCCGCAATTCGTAAAGGTGGGTTCTATTTCCGAACATGGCTTCACGCCACTACCAGGCTATAGATCGAAGTTACGCCTTTGGGCTCAGAGTCGTCGCTTCGCCCACGAGTGTTGTCTGGTCGGAGTCATCGACCTGATAGAATGACCTCCACCGTCGGGCTCGTGGGTTTCCCATGTAAATCAGAGTGCACGCGAGGTAAAGAGCACTTTCCTCAGAAAAAACGCTGTTAATCGTTTGAATGCTTCAAGTGCGGGAAATTCCTCGTGAGCCGTGCATTTTGGTCAGTTAACGCGGCTGTCGGCTTGGG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCAAGUAGCAAUAGUGGGCGUGAACACUGCAUGAAGCUCAGGGAAUGGUUCCUGGAGUUUUAUUUUAGCACACACGAUAAGGUUAAUAGACCUCACUGUCAAGUUCGAUCGUCGAGAAGUGAAGCAAAAACGGUAUCCUCGAGUGGGAAGCUCGGCACUCGCCCCUUGCACGGCGGGAAACCCACGAACACUUGUUUUCCAUAUGCGGCACAGACUUAACCUUAGCGUGUGAGUCGCUUCAGUACAAGAUCAUCGAGACAGUGCACACUGGAUUAAACGUCAACACAUGUGCCGACACUACCUUGGCAGUCUGUAUGGCGAACCAUCGUUCAAAACGGGGACUAAACGGUCAAGGCAGUAGAUGGUUACGAGAGUGCUAGGAUCGUGGCCAAUUCCCACACAUACUUGCGAGUUUGGUUUUUCACACUCGUGUGUAGAAUGGGAUACGCAACCAAAAUCUACCUGCAUACGCGCCUCAUCUCGAUUGUCGAAAUGGAGUGUUUGAAGGUACGUUUGCCAGCGACCUUUGACAGAGUUAUCGCCCUAAAGUCCUCAUACCACGAGCCAGACUCACACUCCAAGAAUGCCUUCGUGGCGCUGCCGCCCGCAAUUCGUAAAGGUGGGUUCUAUUUCCGAACAUGGCUUCACGCCACUACCAGGCUAUAGAUCGAAGUUACGCCUUUGGGCUCAGAGUCGUCGCUUCGCCCACGAGUGUUGUCUGGUCGGAGUCAUCGACCUGAUAGAAUGACCUCCACCGUCGGGCUCGUGGGUUUCCCAUGUAAAUCAGAGUGCACGCGAGGUAAAGAGCACUUUCCUCAGAAAAAACGCUGUUAAUCGUUUGAAUGCUUCAAGUGCGGGAAAUUCCUCGUGAGCCGUGCAUUUUGGUCAGUUAACGCGGCUGUCGGCUUGGG\n"
     ]
    }
   ],
   "source": [
    "# sample dataset: GATGGAACTTGACTACGTAAATT\n",
    "# sample output: GAUGGAACUUGACUACGUAAAUU\n",
    "\n",
    "# take an RNA string formed from the alphabet containing 'A', 'C', 'G' and 'U'\n",
    "# return the transrcibed DNA sequence (as a 'u' string, that has 'A', 'C', 'G' and 'U' nucleotide bases instead of 'A', 'C', 'G' and 'T')\n",
    "\n",
    "##VARIABLES\n",
    "# insert the DNA sequence (as a 't' string)\n",
    "#t = 'GATGGAACTTGACTACGTAAATT'\n",
    "t = str(input('insert the DNA sequence: '))\n",
    "# accumulator varaible for a transcribed RNA string\n",
    "u = ''\n",
    "\n",
    "##PROGRAMME\n",
    "# for each base\n",
    "for base in t:\n",
    "    # if base is 'A' or 'C' or 'G'\n",
    "    if base == 'A' or base == 'C' or base == 'G':\n",
    "        # u becomes u + base\n",
    "        u = u + base\n",
    "    # otherwise\n",
    "    else:\n",
    "        # u becomes u + 'U'\n",
    "        u = u + 'U'\n",
    "\n",
    "# print the RNA sequence\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c566848-06d8-4e68-b6de-87a1120c0501",
   "metadata": {},
   "source": [
    "## REVC - Complementing a Strand of DNA (Loop)\n",
    "https://rosalind.info/problems/revc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f96c4cf-2c0a-45a5-8d31-f590ad99b35f",
   "metadata": {},
   "source": [
    "### Theory\n",
    "The primary structure of a nucleic acid is determined by the ordering of its nucleobases along the sugar-phosphate backbone that constitutes the bonds of the nucleic acid polymer.  Yet primary structure tells us nothing about the larger, 3-dimensional shape of the molecule, which is vital for a complete understanding of nucleic acids.\n",
    "\n",
    "Watson and Crick proposed the following structure for DNA: \n",
    "1. The DNA molecule is made up of two strands, running in opposite directions.\n",
    "2. Each base bonds to a base in the opposite strand. Adenine always bonds with thymine, and cytosine always bonds with guanine; the complement of a base is the base to which it always bonds.\n",
    "3. The two strands are twisted together into a long spiral staircase structure called a double helix.\n",
    "\n",
    "Because they dictate how bases from different strands interact with each other, (1) and (2) above compose the secondary structure of DNA. (3) describes the 3-dimensional shape of the DNA molecule, or its tertiary structure.\n",
    "\n",
    "In light of Watson and Crick's model, the bonding of two complementary bases is called a base pair (bp). Therefore, the length of a DNA molecule will commonly be given in bp instead of nt. By complementarity, once we know the order of bases on one strand, we can immediately deduce the sequence of bases in the complementary strand. These bases will run in the opposite order to match the fact that the two strands of DNA run in opposite directions.\n",
    "\n",
    "\n",
    "The reverse complement of a DNA string is formed by reversing the string and taking the complement of each symbol.\n",
    "\n",
    "We must reverse the string in addition to taking complements because of the directionality of DNA: DNA replication and transcription occurs from the 3' end to the 5' end, and the 3' end of one strand is opposite from the 5' end of the complementary strand. Thus, if we were to simply take complements, then we would be reading the second strand in the wrong direction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5dc71244-290a-4d1a-83b1-66243b2604db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a DNA sequence: CCTCAATGTAGTCCTCGTTAGTGAACCATTCAACCAGACATTCGGACTACTGCGGTACCGTCAGTCTTCAGTTGAAGCAAAGGTACCTACGGGCCGGATTCACTTGTTGATAATACCGATAGTGCTCTGGAAACCAGGCCTAGCGTTCCCGAAGTATTCGTTAGGTCTAGCGGTAGACCCGGCAGGCCAGGAGGAACTGGAAAATTGTACGGTAAGCCAGCTGTATACCGCTAGGTTACCGTGCATCTAGCTACGAGACTATGTCAACATCGTCGTCGCCGATACACCTCATCGCAATTGTATCAGGGTAATTTACGGCATTAACGAGGAAAAAGTACCGAAATGGCGCGCGCCGGTACCAAGTGTTTGGACGGTCTATTCACTGACCGTAACGTGCGACGCTGTTGAGGCAAGGACGCGGCGAATATAATTCGGACGCGGAGGTTGAGAGTGCCATGTCAAACACCACAAGGACCAATCAACTTGGGCAGTCCCTCTGGGTATTAACGGAATGTGTAGATACTACGTTGCACAGTTTTAATCCAATGGTACCATGACGATAGAAAAAGCTAATTTCACACTCTGGGGACAGTGATGCGGTTTGCGCCTTCGCACGTAACATGACTGTAGCCACTTGGGCTTGCCAGAGATGCCTCATTGTTTAATCCTCGGAAACGCTCAAAGGGCAATTCTGACATCTCGATGTTCACTAAACGCGTCACCTTCAAATTAAGACCTGAATTAAACTGATGCCCGGATAGTGCGATTACGTCACCAGTTACACGATATGAGACGGAACCTTCCGTCCTTCTTGTCGCTGTGACCTGACGTACATGAGCTGGGATGAACTGCTTCGAAATGGAGGACCTCACTAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTAGTGAGGTCCTCCATTTCGAAGCAGTTCATCCCAGCTCATGTACGTCAGGTCACAGCGACAAGAAGGACGGAAGGTTCCGTCTCATATCGTGTAACTGGTGACGTAATCGCACTATCCGGGCATCAGTTTAATTCAGGTCTTAATTTGAAGGTGACGCGTTTAGTGAACATCGAGATGTCAGAATTGCCCTTTGAGCGTTTCCGAGGATTAAACAATGAGGCATCTCTGGCAAGCCCAAGTGGCTACAGTCATGTTACGTGCGAAGGCGCAAACCGCATCACTGTCCCCAGAGTGTGAAATTAGCTTTTTCTATCGTCATGGTACCATTGGATTAAAACTGTGCAACGTAGTATCTACACATTCCGTTAATACCCAGAGGGACTGCCCAAGTTGATTGGTCCTTGTGGTGTTTGACATGGCACTCTCAACCTCCGCGTCCGAATTATATTCGCCGCGTCCTTGCCTCAACAGCGTCGCACGTTACGGTCAGTGAATAGACCGTCCAAACACTTGGTACCGGCGCGCGCCATTTCGGTACTTTTTCCTCGTTAATGCCGTAAATTACCCTGATACAATTGCGATGAGGTGTATCGGCGACGACGATGTTGACATAGTCTCGTAGCTAGATGCACGGTAACCTAGCGGTATACAGCTGGCTTACCGTACAATTTTCCAGTTCCTCCTGGCCTGCCGGGTCTACCGCTAGACCTAACGAATACTTCGGGAACGCTAGGCCTGGTTTCCAGAGCACTATCGGTATTATCAACAAGTGAATCCGGCCCGTAGGTACCTTTGCTTCAACTGAAGACTGACGGTACCGCAGTAGTCCGAATGTCTGGTTGAATGGTTCACTAACGAGGACTACATTGAGG\n"
     ]
    }
   ],
   "source": [
    "# sample dataset: AAAACCCGGT\n",
    "# sample output: ACCGGGTTTT\n",
    "\n",
    "# take a DNA string s of length at most 1000 bp\n",
    "# return the reverse complement sc of s\n",
    "\n",
    "##VARIABLES\n",
    "# input DNA sequence\n",
    "#s = 'AAAACCCGGT'\n",
    "# insert DNA sequence\n",
    "s = str(input('insert a DNA sequence:'))\n",
    "# create accumulator variable for reverse complement of a DNA string\n",
    "sc = ''\n",
    "\n",
    "##PROGRAMME\n",
    "# for each base in reversed DNA sequence string\n",
    "for base in s[::-1]:\n",
    "    # if base is 'A'\n",
    "    if base == 'A':\n",
    "        # sc becomes sc + 'T'\n",
    "        sc = sc + 'T'\n",
    "    # or if base is 'T'\n",
    "    elif base == 'T':\n",
    "        # sc becomes sc + 'A'\n",
    "        sc = sc + 'A'\n",
    "    # or if base is 'C'\n",
    "    elif base == 'C':\n",
    "        # sc becomes sc + 'G'\n",
    "        sc = sc + 'G'\n",
    "    # otherwise\n",
    "    else:\n",
    "        # sc becomes sc + 'C'\n",
    "        sc = sc + 'C'\n",
    "\n",
    "# print complementary DNA string\n",
    "print()\n",
    "print('the complementary DNA string, sc:', sc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c91ef029-c18b-40d1-8e76-f9752b64e239",
   "metadata": {},
   "source": [
    "## REVC - Complementing a Strand of DNA (Method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "025d1132-4ef8-4953-9cb5-b7372bb684bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a DNA sequence: AATAGGGCTGAGGCCTACTAGCATTGAATTTTTATGCATATCATCGCCTTACCGCTGAAGTCACAAGTAGATCAAGTGGAAATCTGAACACGTAACAAAAACATGGCCGTATTCTCGTTTGTTTCATACTGCGGAGGTATTCTATTCTGGCGTGGTTTGACATCGCGGTTGCCGGCGGACAAGTATGTCTTTACAGCCGTTTGCCTAGCCCCAGTGATAGAAGACGTGACCTCACCTTTAGCGCAGGTAACAACTTTTGAAAAAAGCTTCTATGGCATCGTATCGTCTTCCGAAAATACGAAGACGTTGGAAACGGATTCATTCGACAACGGCAACATGGATATTGATGAGTATTGACCATCTTGTAAATATAGAGCTCGTCATCTTGATGCATAGTCGTACCGGATGCACCGACGGCAAGGGGCCGTCCTTTCAATACGTTTAGTGCTGCGCAGAAGCCATTGTCTGCTATATGCTACCACCCTCATGCGGCCCAGCGGCATTGCCGACTTTGCTCTGCAAATCACTTGCCGGGAGCAGGTCGCCGAAACCAAGACAGTAGAGCCGGATAACTGTATATTCGAGCACCACTTCTCAATAGGAATGGTAAGCGTCGAATTCCCCAAGAAAAATGTGACTCAATGACTGGCGTGGATAACTAGGGCCTTGGCACCAGGCAAGCCTCCACGCTACGTTTTAGCGGAGGGATCCGCACAATATGGCCATAGGTAACCAGCGATTTGTCATCGGACCTTTCCTTAACGTGAACCGCTGGGTGAGGTAGTTCATCCACACTTTAATCAAACCGCTTGGATGAGCTGACTGGTTCCTGCAACTTTGGGCAGAGAGCACTCCTGATCTAGTATTTACGCTCTCCGGTGGCAGGCTGTCAGCTGATGGGGATCCAACAGCACGCTGTAAGAAGCGCGGGAGAGGGGCCGAGTATCAAATAATCTTCTATG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "the complementary DNA string, sc: CATAGAAGATTATTTGATACTCGGCCCCTCTCCCGCGCTTCTTACAGCGTGCTGTTGGATCCCCATCAGCTGACAGCCTGCCACCGGAGAGCGTAAATACTAGATCAGGAGTGCTCTCTGCCCAAAGTTGCAGGAACCAGTCAGCTCATCCAAGCGGTTTGATTAAAGTGTGGATGAACTACCTCACCCAGCGGTTCACGTTAAGGAAAGGTCCGATGACAAATCGCTGGTTACCTATGGCCATATTGTGCGGATCCCTCCGCTAAAACGTAGCGTGGAGGCTTGCCTGGTGCCAAGGCCCTAGTTATCCACGCCAGTCATTGAGTCACATTTTTCTTGGGGAATTCGACGCTTACCATTCCTATTGAGAAGTGGTGCTCGAATATACAGTTATCCGGCTCTACTGTCTTGGTTTCGGCGACCTGCTCCCGGCAAGTGATTTGCAGAGCAAAGTCGGCAATGCCGCTGGGCCGCATGAGGGTGGTAGCATATAGCAGACAATGGCTTCTGCGCAGCACTAAACGTATTGAAAGGACGGCCCCTTGCCGTCGGTGCATCCGGTACGACTATGCATCAAGATGACGAGCTCTATATTTACAAGATGGTCAATACTCATCAATATCCATGTTGCCGTTGTCGAATGAATCCGTTTCCAACGTCTTCGTATTTTCGGAAGACGATACGATGCCATAGAAGCTTTTTTCAAAAGTTGTTACCTGCGCTAAAGGTGAGGTCACGTCTTCTATCACTGGGGCTAGGCAAACGGCTGTAAAGACATACTTGTCCGCCGGCAACCGCGATGTCAAACCACGCCAGAATAGAATACCTCCGCAGTATGAAACAAACGAGAATACGGCCATGTTTTTGTTACGTGTTCAGATTTCCACTTGATCTACTTGTGACTTCAGCGGTAAGGCGATGATATGCATAAAAATTCAATGCTAGTAGGCCTCAGCCCTATT\n"
     ]
    }
   ],
   "source": [
    "# sample dataset: AAAACCCGGT\n",
    "# sample output: ACCGGGTTTT\n",
    "\n",
    "# take a DNA string s of length at most 1000 bp\n",
    "# return the reverse complement sc of s\n",
    "\n",
    "##VARIABLES\n",
    "# input DNA sequence\n",
    "#s = 'AAAACCCGGT'\n",
    "# insert DNA sequence\n",
    "s = str(input('insert a DNA sequence:'))\n",
    "# reverse the inserted DNA sequence\n",
    "rs = s[::-1]\n",
    "# transform into lower case letters\n",
    "rs = rs.lower()\n",
    "\n",
    "# in one go\n",
    "#s = str(input('insert a DNA sequence:'))[::-1].lower()\n",
    "\n",
    "##PROGRAMME\n",
    "# raplace each lowe case letter with an upper case letter of a complementary nucleobase\n",
    "sc = rs.replace('a', 'T') # firstly use rs and insert result into sc\n",
    "sc = sc.replace('t', 'A') # from this point onward work with sc only\n",
    "sc = sc.replace('c', 'G')\n",
    "sc = sc.replace('g', 'C')\n",
    "\n",
    "# print complementary DNA string\n",
    "print()\n",
    "print('the complementary DNA string, sc:', sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6135267b-d416-47cc-9573-093df353fc20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method descriptor replace:\n",
      "\n",
      "replace(self, old, new, /, count=-1) unbound builtins.str method\n",
      "    Return a copy with all occurrences of substring old replaced by new.\n",
      "\n",
      "      count\n",
      "        Maximum number of occurrences to replace.\n",
      "        -1 (the default value) means replace all occurrences.\n",
      "\n",
      "    If the optional argument count is given, only the first count occurrences are\n",
      "    replaced.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(str.replace)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a83862bb-072f-49d2-bb4d-a69cb458c0dd",
   "metadata": {},
   "source": [
    "## HAMM - Counting Point Mutations (Loop/Con)\n",
    "https://rosalind.info/problems/hamm/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a6a8c0-c6fd-4f05-9e3b-28dc71067947",
   "metadata": {},
   "source": [
    "### Theory\n",
    "\n",
    "A mutation is a mistake which occurs during the creation/copying of a nucleic acid, in particular DNA. Sum of micro-mutations over many generations might lead to positive macro effects.\n",
    "\n",
    "The simpliest and most common type of nucleic acid mutation is a point mutation, which replaces one base with another a a single nucleotide. In the case of DNA, a point mutation must change the complementary base accordingly.\n",
    "\n",
    "Two DNA strands taken from different organism or species genomes are homologous if they share a recent ancestor; thus, counting the number of bases at which homologous strands differ provides us with the minimum number of point mutations that could have occurred on the evolutionary path between the two strands. Minimising the number of point mutations separating two species needs to be done, since the principle of parsimony demands that evolutionary histories should be be as simply explained as possible.\n",
    "\n",
    "\n",
    "The Hamming distance between two strings having the same length is the minimum number of symbol substitutions required to transform one string into the other. If the strings are given by s1 and s2, then we write the Hamming distance between them as dH(s1,s2).\n",
    "\n",
    "We can compute the Hamming distance by visual inspection: the Hamming distance between two strings is simply the number of positions in the strings at which corresponding symbols differ. For example, the Hamming distance between the two strings in the figure below must be equal to 7."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "cfd93272-9346-44d3-b076-c950f532a3b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "first DNA sequence: TAATGTGCAGGTTACGTACATGGCTTACACTGGGCGAAAGACTGTTGCAATGCGTTCCTCCCTCGCAGACTCCTACTACGTGTACCTGGTAGCAGGGATCCGGGGAGGACTAGACGCCTGTCTCAATAACGGATAGGAGCGGCGCTATAATCGCGCGCATTTAGTTGAGAGACGCCCTCACCCGTACTAAACTCGGAAGTCCGGGATTCTAAGCCAGAGATTCGGGTGCTCCTACTTTATCGAATTGTGTAAGATCTAAAGATTGATTCGTCCGCGGAGGTCTCCCAGCCCAGTATTATAGCGTCAGTTTTTCACATGTCTGAGCCTAACAACGGCAAGCACGCCAGAAGAATACCTGCGTCCATCGACCGGAATAGCTACTGAACGTTGAGGTGACACGACCGGAAGACGTTTACCTCGGCTACTACATCTTTGACGAAAGACAATTGTAGCTTTCACGGTGCTTCTTACTATTCGAACGAATTCGTTCTAATACTTAAATTCAATTCACGTGTCGCCTTCAAGTAGTCACTAACTCGGCCTCTTGGTACTCCACGGTGTATTATAGGTGGCCTTTGACTCCTGGACGCTGGTTGGTCCAGCTCTGCGCTACTTGGTTCCCAGTTGAACGGCGCGGCGTCATACACGTCATCTAAGCGAGCTCGGCCGCACATTCCCGTTATGCGGACATCGGGTACTTAAATGCGCTTAGCCTGGGGTGTGTTTCAATGCAAGTAAGCTCCACGCCACAACGTTGACCCCCCAGGTACCTGAGACCGCAGCTTGTTGTCAGAGTATCATGAATCCGGCGCAAGAGGCCGCTACACAAGGCCGAAGGTCTTTGATGATCCTTCAATACTGTCCTCATTAAACCAACTTCGAAAAAGATTTCAGGGCATGAGCAACCCCTGAAAACCATATAATACGAACTGCTGCTGTGTGGTGACA\n",
      "second DNA sequence: ACCTGTGCGTGTGTAGTTCGTAGCGGAGACAGTGCCATGGATCATGGCAGCACGTTTCACTAGGTCAAAGTTCGATGTAGGCCACCTCCGGTGATCCATCACGGGCAGTCTACGCTCTTTTGTCAATACCATTGCAGTAGGACTCTTTCACATACCACTAATACTGGTGAGACGAGCCGGAGCGTAGTAGTTTCGCAGAGCCAGTTTAAAAATCACTGGATGCCGTAACTTCTACAATATAAGAAAGTCATAGCGTCGCACTACGATTATTTCGCGGTGATCTCAGAGCCTACTGCCAAGCTATTTGTTTTCAACGTGTGCCCACAGTTCTAATACAATGTTGAAAGACCGGGTAGTTGTCGCACCGGCTGGAACCCCCATTCAGAGGTGCGGCAACAAGACCGGAGCTCGGCTTCGCCGGTCTGTTCTTCTCCTCCCCCGGGGAAATGCCGCTTTCACGCTCCGTGGGCCTTTTGCTTCGTATTCTACTATACACCTACATGGAAGACATGCGCTGCCTAGAATCAACCAAGAGCTCGGGGGCGGGGTAATCGAAGGACTAGTATGGACGGCTCTAATCCCCTGGATGGCCGCTAGCCCCACGTTATTCCTAATTTACGCCCTTGTATAGTTGTGTCAACACAACCGGCCTCAGGCCTCCCGATCCGGGAGCAAACGGATACTAGGCGCCCAAGGACTAAATGGTCATTGGCCTGCGCTTTGCGTTCATGTTCTGGAACGCCGGCTCTTAACGATTATGCCGCAGGTTGCATCCAGCACGGTGTGATAACGGAATACTATGCATACGGTGTCATACAGTGCTACAGACGGAAGCAGGGCTTTGATCATACTCCTATTCAGTTTCACATAGCACAATATATACCGCAATGAATGTGAGAGAAATACGCCCCAGGAGCGCACGTGACGTATAGTTGGACTGCGTTGGCG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Hamming distance is: 493\n"
     ]
    }
   ],
   "source": [
    "# return the hamming distance (number of mismatching nucleotide bases)\n",
    "\n",
    "##VARIABLE\n",
    "#s = 'GAGCCTACTAACGGGAT'\n",
    "s = str(input('first DNA sequence:'))\n",
    "#t = 'CATCGTAATGACGGCCT'\n",
    "t = str(input('second DNA sequence:'))\n",
    "d = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for particular position (index) in range of length of the string, do\n",
    "for position in range(len(s)): # alternatively a while loop\n",
    "    # if nucleotide base at the particular position in 's' string is not the same as the nucleotide base at the aprticular position in 't' string, then do\n",
    "    if s[position] != t[position]:\n",
    "        # add 1 to hamming distance\n",
    "        d = d + 1 # alternatively +=\n",
    "    #if not, then continue to with another iteration\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "# print Hamming distance\n",
    "print()\n",
    "print('Hamming distance is:', d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d338f783-477a-4528-8bba-a18b738bb869",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
