{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "95d86f75-fdcd-40c8-a5a7-93bcf717702e",
   "metadata": {},
   "source": [
    "# Jupyter Notebook for Rosalinda INI Exercises"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25fbe80a-f2ae-45e9-8405-78d2dc5c2e7b",
   "metadata": {},
   "source": [
    "## INI 1 - Installing Python (Import)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b20cf8a-7bc5-4b9d-a143-8ffd9e97ba6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Zen of Python, by Tim Peters\n",
      "\n",
      "Beautiful is better than ugly.\n",
      "Explicit is better than implicit.\n",
      "Simple is better than complex.\n",
      "Complex is better than complicated.\n",
      "Flat is better than nested.\n",
      "Sparse is better than dense.\n",
      "Readability counts.\n",
      "Special cases aren't special enough to break the rules.\n",
      "Although practicality beats purity.\n",
      "Errors should never pass silently.\n",
      "Unless explicitly silenced.\n",
      "In the face of ambiguity, refuse the temptation to guess.\n",
      "There should be one-- and preferably only one --obvious way to do it.\n",
      "Although that way may not be obvious at first unless you're Dutch.\n",
      "Now is better than never.\n",
      "Although never is often better than *right* now.\n",
      "If the implementation is hard to explain, it's a bad idea.\n",
      "If the implementation is easy to explain, it may be a good idea.\n",
      "Namespaces are one honking great idea -- let's do more of those!\n"
     ]
    }
   ],
   "source": [
    "# display the Zen of Python\n",
    "import this"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73391690-d4cc-4ebe-b60c-365011a54038",
   "metadata": {},
   "source": [
    "## INI 2 - Variables and Some Arithmetic (Mathematical Operation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "12f95e7b-e3fc-4867-af0e-738b5a006ad8",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "Interrupted by user",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 4\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# calculation of the size of a square on the hypotenuse (right-angled triangle)\u001b[39;00m\n\u001b[32m      2\u001b[39m \n\u001b[32m      3\u001b[39m \u001b[38;5;66;03m# input side lengths as variables\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m4\u001b[39m a = \u001b[38;5;28mint\u001b[39m(\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43msize of \u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43ma\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m side: \u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m) \u001b[38;5;66;03m# size of 'a' side\u001b[39;00m\n\u001b[32m      5\u001b[39m b = \u001b[38;5;28mint\u001b[39m(\u001b[38;5;28minput\u001b[39m(\u001b[33m'\u001b[39m\u001b[33msize of \u001b[39m\u001b[33m\"\u001b[39m\u001b[33mb\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m side: \u001b[39m\u001b[33m'\u001b[39m)) \u001b[38;5;66;03m# size of 'b' side\u001b[39;00m\n\u001b[32m      7\u001b[39m \u001b[38;5;66;03m# calcualate (using exponentiation)\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/carpentries/lib/python3.14/site-packages/ipykernel/kernelbase.py:1396\u001b[39m, in \u001b[36mKernel.raw_input\u001b[39m\u001b[34m(self, prompt)\u001b[39m\n\u001b[32m   1394\u001b[39m     msg = \u001b[33m\"\u001b[39m\u001b[33mraw_input was called, but this frontend does not support input requests.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1395\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m StdinNotImplementedError(msg)\n\u001b[32m-> \u001b[39m\u001b[32m1396\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_input_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1397\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1398\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_get_shell_context_var\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_shell_parent_ident\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1399\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mget_parent\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mshell\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1400\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpassword\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m   1401\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniforge3/envs/carpentries/lib/python3.14/site-packages/ipykernel/kernelbase.py:1441\u001b[39m, in \u001b[36mKernel._input_request\u001b[39m\u001b[34m(self, prompt, ident, parent, password)\u001b[39m\n\u001b[32m   1438\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[32m   1439\u001b[39m     \u001b[38;5;66;03m# re-raise KeyboardInterrupt, to truncate traceback\u001b[39;00m\n\u001b[32m   1440\u001b[39m     msg = \u001b[33m\"\u001b[39m\u001b[33mInterrupted by user\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m-> \u001b[39m\u001b[32m1441\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m(msg) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m   1442\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[32m   1443\u001b[39m     \u001b[38;5;28mself\u001b[39m.log.warning(\u001b[33m\"\u001b[39m\u001b[33mInvalid Message:\u001b[39m\u001b[33m\"\u001b[39m, exc_info=\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: Interrupted by user"
     ]
    }
   ],
   "source": [
    "# calculation of the size of a square on the hypotenuse (right-angled triangle)\n",
    "\n",
    "# input side lengths as variables\n",
    "a = int(input('size of \"a\" side: ')) # size of 'a' side\n",
    "b = int(input('size of \"b\" side: ')) # size of 'b' side\n",
    "\n",
    "# calcualate (using exponentiation)\n",
    "square = (a**2 + b**2)\n",
    "\n",
    "# print the results (of calculation)\n",
    "print('square on the hypotenuse:', square)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3315bfd7-8abf-411c-b5b6-e3cdf3d52780",
   "metadata": {},
   "source": [
    "## INI 3 - Strings and Lists (Slice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63002c2d-997e-41c5-ad04-1cec318cdd34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# slicing of a provided string based on determined positions to be sliced, not excluding, but comprising \"b\" and \"d\" indices.\n",
    "\n",
    "# input string (to be sliced)\n",
    "s = str(input('provide a string: '))\n",
    "\n",
    "# input indices (to determine the extent of slicing) \n",
    "a = int(input('provide an index \"a\": '))\n",
    "b = int(input('provide an index \"b\": '))\n",
    "c = int(input('provide an index \"c\": '))\n",
    "d = int(input('provide an index \"d\": '))\n",
    "\n",
    "# print results (of slicing)\n",
    "print(s[a:b+1], s[c:d+1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b877fd4-9698-44d2-841f-b5213a0f3107",
   "metadata": {},
   "source": [
    "## INI 4 - Conditions and Loops (Loop/Con - is in?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "912e84f8-9fff-4633-b4b7-042d72a84cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return the sum of all odd integers from a through b, inclusively.\n",
    "\n",
    "##VARIABLES\n",
    "# input integers (according to the following scheme: a < b < 10000)\n",
    "a = int(input('insert integer value for \"a\": '))\n",
    "b = int(input('insert integer value for \"b\": '))\n",
    "\n",
    "# create a list of odd numbers\n",
    "odd_list = [1, 3, 5, 7, 9]\n",
    "\n",
    "# create a variable with value 0\n",
    "odd_sum = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for loop (from a through b, inclusively)\n",
    "for num in range(a, b + 1):\n",
    "    #print(num)\n",
    "    num_list = list(str(num)) # transform a number into a list\n",
    "    #print(num_list)\n",
    "    last_num = int(num_list[len(num_list) - 1]) # get the last number\n",
    "    #print(last_num)\n",
    "\n",
    "    # if condition (if the last number is odd, add the number to the sum of previous odd numbers)\n",
    "    if last_num in odd_list: #alternatively: a % 2 == 1 <-divided by 2 and if odd, remainder is 1, applies to 1 as well\n",
    "    #if last_num % 2 == 1:\n",
    "        #print('odd')\n",
    "        odd_sum = odd_sum + num\n",
    "    #else:\n",
    "        #print('even')\n",
    "        \n",
    "# print results (of conditional calculation)\n",
    "print(odd_sum)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c35ebbea-945d-4410-bf89-7b6e563f4e69",
   "metadata": {},
   "source": [
    "## INI4 - Conditions and Loops (Loop/Con - is odd/even?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b91b24-67d1-4aa4-8dc5-ec3fa335a064",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return the sum of all odd integers from a through b, inclusively.\n",
    "\n",
    "##VARIABLES\n",
    "# input integers (according to the following scheme: a < b < 10000)\n",
    "a = 100\n",
    "b = 200\n",
    "odd_sum = 0\n",
    "\n",
    "for num in range(a, b + 1):\n",
    "    if num % 2 != 0:\n",
    "        odd_sum = odd_sum + num\n",
    "\n",
    "print(odd_sum)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "767e0d04-544b-47cb-b695-fb836a9ce150",
   "metadata": {},
   "source": [
    "## INI 5 - Working with Files (Slice)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "557be151-ec4d-4001-aeb2-f3cd1f823f0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Some things in life are bad, they can really make you mad\n",
      "Other things just make you swear and curse\n",
      "When you're chewing on life's gristle, don't grumble give a whistle\n",
      "This will help things turn out for the best\n",
      "Always look on the bright side of life\n",
      "Always look on the right side of life\n",
      "If life seems jolly rotten, there's something you've forgotten\n",
      "And that's to laugh and smile and dance and sing\n",
      "When you're feeling in the dumps, don't be silly, chumps\n",
      "Just purse your lips and whistle, that's the thing\n",
      "So, always look on the bright side of death\n",
      "Just before you draw your terminal breath\n",
      "Life's a counterfeit and when you look at it\n",
      "Life's a laugh and death's the joke, it's true\n",
      "You see, it's all a show, keep them laughing as you go\n",
      "Just remember the last laugh is on you\n",
      "Always look on the bright side of life\n",
      "And always look on the right side of life\n",
      "Always look on the bright side of life\n",
      "And always look on the right side of life\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# return a file containing all the even-numbered lines form the original file and assume 1-based numbering of lines\n",
    "\n",
    "# input string: 'Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat'\n",
    "\n",
    "# create an example file\n",
    "# file_in = open('../../rosalind_INI4.txt', 'w')\n",
    "# file_in.write('Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat')\n",
    "# file_in.close()\n",
    "\n",
    "##VARIABLES\n",
    "# input file path\n",
    "#input_path = str(input('insert file path for input text file:'))\n",
    "input_path = '../../rosalind_INI5.txt'\n",
    "# output file path\n",
    "#output_path = str(input('insert file path for output text file:'))\n",
    "output_path = '../../rosalind_INI5_output.txt'\n",
    "\n",
    "# load text\n",
    "file_in = open(input_path, 'r')\n",
    "# create output file\n",
    "file_out = open(output_path, 'w')\n",
    "\n",
    "##PROGRAMME\n",
    "# extract the number of lines\n",
    "#line_count = len(file_in.readlines()) + 1\n",
    "# reload input file\n",
    "#file_in = open(input_path, 'r')\n",
    "# extract even-numbered lines (odd numbers in slicing, because starts from 0)\n",
    "output = file_in.readlines()[1::2] \n",
    "\n",
    "##NOTE\n",
    "# if length is defined ([1:line_count:2]), after line_count, input file needs to be reloaded\n",
    "\n",
    "# convert list of strings into a unified string\n",
    "string = ''.join(output)\n",
    "# write output into the output file\n",
    "file_out.write(string)\n",
    "# close file modifications\n",
    "file_out.close()\n",
    "\n",
    "# read output file\n",
    "file_out = open(output_path, 'r')\n",
    "# transform list into a string\n",
    "result = ''.join(file_out.readlines())\n",
    "# print results\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d275f627-3293-498e-9138-586fbf8f3dac",
   "metadata": {},
   "source": [
    "## INI 5 - Working with Files (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18ee6d31-14bf-4154-87a4-33d8cfed8958",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return a file containing all the even-numbered lines form the original file and assume 1-based numbering of lines\n",
    "\n",
    "# input string: 'Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat'\n",
    "\n",
    "# create an example file\n",
    "# file_in = open('../../rosalind_INI4.txt', 'w')\n",
    "# file_in.write('Bravely bold Sir Robin rode forth from Camelot\\nYes, brave Sir Robin turned about\\nHe was not afraid to die, O brave Sir Robin\\nAnd gallantly he chickened out\\nHe was not at all afraid to be killed in nasty ways\\nBravely talking to his feet\\nBrave, brave, brave, brave Sir Robin\\nHe beat a very brave retreat')\n",
    "# file_in.close()\n",
    "\n",
    "##VARIABLES\n",
    "# input file path\n",
    "#input_path = str(input('insert file path for input text file:'))\n",
    "input_path = '../../rosalind_INI5.txt'\n",
    "# output file path\n",
    "#output_path = str(input('insert file path for output text file:'))\n",
    "output_path = '../../rosalind_INI5_output.txt'\n",
    "\n",
    "# load text\n",
    "file_in = open(input_path, 'r')\n",
    "# create output file\n",
    "file_out = open(output_path, 'w')\n",
    "# create an accumulator variable\n",
    "position = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for each line in text file\n",
    "for line in file_in:\n",
    "    # if position is odd\n",
    "    if position % 2 == 1:\n",
    "        # write the line into output file\n",
    "        file_out.write(line)\n",
    "        # add 1 to position\n",
    "        position += 1\n",
    "    # otherwise - does not need to be included, if + add 1 would be enough\n",
    "    else:\n",
    "        # add 1 to position \n",
    "        position += 1\n",
    "        # continue with another iteration\n",
    "        continue\n",
    "\n",
    "# close writing into the output file\n",
    "file_out.close()\n",
    "\n",
    "# read output file\n",
    "file_out = open(output_path, 'r')\n",
    "# transform list into a string\n",
    "result = ''.join(file_out.readlines())\n",
    "# print results\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53bf4e1a-7f1c-4cb1-bb1a-c50463a82f10",
   "metadata": {},
   "source": [
    "## INI 6 - Dictionaries (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc1a062-00a1-421e-a4d5-c96c5cb01f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return the number of occurrences of each word in variable 's' (string), where words are separated by spaces; words are case-sensitive; lines in the output can be in any order\n",
    "\n",
    "## VARIABLES\n",
    "# input string and dictionary-based accumulator variable\n",
    "#s = 'We tried list and we tried dicts also we tried Zen'\n",
    "s = str(input('insert string:'))\n",
    "word_count = {}\n",
    "\n",
    "## PROGRAMME - extracting information\n",
    "# for loop (from the first word from the list to the last one)\n",
    "for word in s.split(): # hint: use (' ') for space or other delimiter character\n",
    "    #if word exists in the dictionary, then add 1 to its value\n",
    "    if word in word_count:\n",
    "        word_count[word] = word_count[word] + 1 # alternatively: +=\n",
    "        # alternatively\n",
    "        #word_count.update({word : word_count[word] + 1})\n",
    "    #if not, then create new key-value pair with a value 1\n",
    "    else:\n",
    "        word_count[word] = 1\n",
    "        # alternatively\n",
    "        #word_count.update({word : 1})\n",
    "\n",
    "#print(word_count) # debugging\n",
    "#word_count = dict(sorted(word_count.items())) # alphabetically ordered dictionary\n",
    "\n",
    "## VARIABLES\n",
    "# integer-based accumulator variable\n",
    "index = 0\n",
    "\n",
    "## PROGRAMME - processing information\n",
    "# for loop (from the first key-value pair from the dictionary to the last one)\n",
    "for word in word_count: \n",
    "    # print words and their occurence number\n",
    "    print(list(word_count)[index], word_count[word])\n",
    "    # add one to the accumulator variable\n",
    "    index = index + 1\n",
    "\n",
    "# hint: alternative easy way\n",
    "# for word, value in word_count.items():\n",
    "#     print(word, value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b353b205-3799-445a-b19b-5d3acef2d702",
   "metadata": {},
   "source": [
    "## DNA - Counting Nucleotides (Method; Loop/Con; Dictionary/Method)\n",
    "https://rosalind.info/problems/dna/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75546f1a-b5e4-46d0-9b82-1ffdddc9301e",
   "metadata": {},
   "source": [
    "### Theory\n",
    "Considering the life and living organic material, the cell is recognised as the building block of life. The centre of the activity is considered to be in the nucles (only in eukaryotic cells, where it is present). The nucleus is filled with macromolecules called chromatin. During eukaryotic cell division (mitosis), most of the chromatin condenses into long, thin strings called chromosomes.\n",
    "\n",
    "One class of the macromolecules contained in chromatin are called nucleic acids. Nucleic acids are polymers, or repeating chains of smaller, similarly structured molecules known as monomers. Because of their tendency to be long and thin, nucleic acid polymers are commonly called strands.\n",
    "\n",
    "The nucleic acid monomer is called a nucleotide and is used as a unit of strand length (abbreviated to nt). Each nucleotide is formed of three parts: a sugar molecule, a negatively charged ion called a phosphate, and a compound called a nucleobase (\"base\" for short). Polymerisation is achieved as the sugar of one nucleotide bonds to the phosphate of the next nucleotide in the chain, which forms a sugar-phosphate backbone for the nucleic acid strand. A key point is that the nucleotides of a specific type of nucleic acid always contain the same sugar and phosphate molecules, and they differ only in their choice of base. Thus, one strand of a nucleic acid can be differentiated from another based solely on the order of its bases; this ordering of bases defines a nucleic acid's primary structure.\n",
    "\n",
    "A strand of deoxyribose nucleic acid (DNA), in which the sugar is called deoxyribose, posseses only four choices for nucleobases, molecules called adenine (A), cytosine (C), guanine (G), and thymine (T).\n",
    "\n",
    "DNA is found in all living organisms on Earth, including bacteria; it is even found in many viruses (which are often considered to be nonliving). Because of its importance, we reserve the term genome to refer to the sum total of the DNA contained in an organism's chromosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22a533c3-340d-4d00-a52f-46544c3c9a3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db7b261c-a701-49c1-84c7-e0793210fcb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert the DNA sequnce (as a string)\n",
    "#s = str(input('insert the DNA sequence: '))\n",
    "\n",
    "# make a (DNA) list with counts per each nucleotide base (\"A\", \"C\", \"G\", \"T\")\n",
    "#DNA = [s.count('A'), s.count('C'), s.count('G'), s.count('T')]\n",
    "\n",
    "# print the counts (for nucleotide base - \"A\", \"C\", \"G\", \"T\")\n",
    "\n",
    "\n",
    "## print('A occurs: ', DNA[0], 'times',\n",
    "##       'C occurs: ', DNA[1], 'times',\n",
    "##       'G occurs: ', DNA[2], 'times',\n",
    "##       'T occurs: ', DNA[3], 'times',)\n",
    "\n",
    "#print(int(DNA[0]), int(DNA[1]), int(DNA[2]), int(DNA[3]))\n",
    "\n",
    "\n",
    "\n",
    "##BONUS - Rosalind DNA, but use a dictionary format\n",
    "\n",
    "# input\n",
    "#AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC\n",
    "#s = str(input('DNA string:'))\n",
    "# s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'\n",
    "\n",
    "# s_dic = {'A': s.count('A'),\n",
    "#          'C': s.count('C'),\n",
    "#          'G': s.count('G'),\n",
    "#          'T': s.count('T')\n",
    "#         }\n",
    "         \n",
    "# print(s_dic)\n",
    "\n",
    "\n",
    "\n",
    "##BONUS - Rosalind DNA, but use a loop\n",
    "\n",
    "##VARIABLES\n",
    "#s = 'AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC'\n",
    "s = str(input('DNA string:'))\n",
    "count = {}\n",
    "##PROGRAMME\n",
    "for nucleotide_base in list(s):\n",
    "    if nucleotide_base in count:\n",
    "        count[nucleotide_base] = count[nucleotide_base] + 1\n",
    "        #count.update({nucleotide_base : count[nucleotide_base] + 1})\n",
    "    else:\n",
    "        count[nucleotide_base] = 1\n",
    "        #count.update({nucleotide_base : 1})\n",
    "\n",
    "count = dict(sorted(count.items())) # alphabetical sorting\n",
    "\n",
    "print(list(count.values())[0], list(count.values())[1], list(count.values())[2], list(count.values())[3]) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61328818-8801-472b-872b-d01ae539064c",
   "metadata": {},
   "source": [
    "## RNA - Transcribing DNA into RNA (Method)\n",
    "https://rosalind.info/problems/rna/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f39fd0e-bfbe-4871-9885-15c4da96f698",
   "metadata": {},
   "source": [
    "### Theory\n",
    "RNA is a second nucleic acid existing alongside DNA in the chromatin and contains a different sugar called ribose, which came to be known as ribose nucleic acid (RNA). RNA differs from DNA in that it contains a base called uracil in place of thymine. RNA is present in the cells of all life forms on Earth.\n",
    "\n",
    "The structure of DNA and RNA is similar because DNA seves as a blueprint for the creation of special sort of RNA, called messenger RNA (mRNA). mRNA is created during RNA transcription, during which a strand of DNA is used as a template for constructing a strand of RNA by copying nucleotides one at a time, where uracil is used in place of thymine.\n",
    "\n",
    "In eukaryotes, DNA remains in the nucleus, while RNA can enter the far reaches of the cell to carry out DNA's instructions.\n",
    "\n",
    "\n",
    "An RNA string is a string that represents the order of nucleobases along a strand of RNA. For this reason, RNA strings are constructed from the alphabet {A, C, G, U}, whose symbols represent the bases adenine, cytosine, guanine, and uracil. For example, \"AAGAUGCCGU\" is an RNA string of length 10."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f7a405f-5ea6-496d-81a1-ec6de82a87ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert the DNA sequence (as a 't' string)\n",
    "t = str(input('insert the DNA sequence: '))\n",
    "\n",
    "\n",
    "# return the transrcibed DNA sequence (as a 'u' string, that has 'A', 'C', 'G' and 'U' nucleotide bases instead of 'A', 'C', 'G' and 'T')\n",
    "u = t.replace('T', 'U')\n",
    "\n",
    "# print the RNA sequence\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e1349a1-b278-44a0-bc38-eee992033e76",
   "metadata": {},
   "source": [
    "## RNA - Transcribing DNA into RNA (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a0f46008-b16e-48cd-9e57-9b760fbc2629",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert the DNA sequence:  AGCAAGTAGCAATAGTGGGCGTGAACACTGCATGAAGCTCAGGGAATGGTTCCTGGAGTTTTATTTTAGCACACACGATAAGGTTAATAGACCTCACTGTCAAGTTCGATCGTCGAGAAGTGAAGCAAAAACGGTATCCTCGAGTGGGAAGCTCGGCACTCGCCCCTTGCACGGCGGGAAACCCACGAACACTTGTTTTCCATATGCGGCACAGACTTAACCTTAGCGTGTGAGTCGCTTCAGTACAAGATCATCGAGACAGTGCACACTGGATTAAACGTCAACACATGTGCCGACACTACCTTGGCAGTCTGTATGGCGAACCATCGTTCAAAACGGGGACTAAACGGTCAAGGCAGTAGATGGTTACGAGAGTGCTAGGATCGTGGCCAATTCCCACACATACTTGCGAGTTTGGTTTTTCACACTCGTGTGTAGAATGGGATACGCAACCAAAATCTACCTGCATACGCGCCTCATCTCGATTGTCGAAATGGAGTGTTTGAAGGTACGTTTGCCAGCGACCTTTGACAGAGTTATCGCCCTAAAGTCCTCATACCACGAGCCAGACTCACACTCCAAGAATGCCTTCGTGGCGCTGCCGCCCGCAATTCGTAAAGGTGGGTTCTATTTCCGAACATGGCTTCACGCCACTACCAGGCTATAGATCGAAGTTACGCCTTTGGGCTCAGAGTCGTCGCTTCGCCCACGAGTGTTGTCTGGTCGGAGTCATCGACCTGATAGAATGACCTCCACCGTCGGGCTCGTGGGTTTCCCATGTAAATCAGAGTGCACGCGAGGTAAAGAGCACTTTCCTCAGAAAAAACGCTGTTAATCGTTTGAATGCTTCAAGTGCGGGAAATTCCTCGTGAGCCGTGCATTTTGGTCAGTTAACGCGGCTGTCGGCTTGGG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCAAGUAGCAAUAGUGGGCGUGAACACUGCAUGAAGCUCAGGGAAUGGUUCCUGGAGUUUUAUUUUAGCACACACGAUAAGGUUAAUAGACCUCACUGUCAAGUUCGAUCGUCGAGAAGUGAAGCAAAAACGGUAUCCUCGAGUGGGAAGCUCGGCACUCGCCCCUUGCACGGCGGGAAACCCACGAACACUUGUUUUCCAUAUGCGGCACAGACUUAACCUUAGCGUGUGAGUCGCUUCAGUACAAGAUCAUCGAGACAGUGCACACUGGAUUAAACGUCAACACAUGUGCCGACACUACCUUGGCAGUCUGUAUGGCGAACCAUCGUUCAAAACGGGGACUAAACGGUCAAGGCAGUAGAUGGUUACGAGAGUGCUAGGAUCGUGGCCAAUUCCCACACAUACUUGCGAGUUUGGUUUUUCACACUCGUGUGUAGAAUGGGAUACGCAACCAAAAUCUACCUGCAUACGCGCCUCAUCUCGAUUGUCGAAAUGGAGUGUUUGAAGGUACGUUUGCCAGCGACCUUUGACAGAGUUAUCGCCCUAAAGUCCUCAUACCACGAGCCAGACUCACACUCCAAGAAUGCCUUCGUGGCGCUGCCGCCCGCAAUUCGUAAAGGUGGGUUCUAUUUCCGAACAUGGCUUCACGCCACUACCAGGCUAUAGAUCGAAGUUACGCCUUUGGGCUCAGAGUCGUCGCUUCGCCCACGAGUGUUGUCUGGUCGGAGUCAUCGACCUGAUAGAAUGACCUCCACCGUCGGGCUCGUGGGUUUCCCAUGUAAAUCAGAGUGCACGCGAGGUAAAGAGCACUUUCCUCAGAAAAAACGCUGUUAAUCGUUUGAAUGCUUCAAGUGCGGGAAAUUCCUCGUGAGCCGUGCAUUUUGGUCAGUUAACGCGGCUGUCGGCUUGGG\n"
     ]
    }
   ],
   "source": [
    "# sample dataset: GATGGAACTTGACTACGTAAATT\n",
    "# sample output: GAUGGAACUUGACUACGUAAAUU\n",
    "\n",
    "# take an RNA string formed from the alphabet containing 'A', 'C', 'G' and 'U'\n",
    "# return the transrcibed DNA sequence (as a 'u' string, that has 'A', 'C', 'G' and 'U' nucleotide bases instead of 'A', 'C', 'G' and 'T')\n",
    "\n",
    "##VARIABLES\n",
    "# insert the DNA sequence (as a 't' string)\n",
    "#t = 'GATGGAACTTGACTACGTAAATT'\n",
    "t = str(input('insert the DNA sequence: '))\n",
    "# accumulator varaible for a transcribed RNA string\n",
    "u = ''\n",
    "\n",
    "##PROGRAMME\n",
    "# for each base\n",
    "for base in t:\n",
    "    # if base is 'A' or 'C' or 'G'\n",
    "    if base == 'A' or base == 'C' or base == 'G':\n",
    "        # u becomes u + base\n",
    "        u = u + base\n",
    "    # otherwise\n",
    "    else:\n",
    "        # u becomes u + 'U'\n",
    "        u = u + 'U'\n",
    "\n",
    "# print the RNA sequence\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c566848-06d8-4e68-b6de-87a1120c0501",
   "metadata": {},
   "source": [
    "## REVC - Complementing a Strand of DNA (Loop)\n",
    "https://rosalind.info/problems/revc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f96c4cf-2c0a-45a5-8d31-f590ad99b35f",
   "metadata": {},
   "source": [
    "### Theory\n",
    "The primary structure of a nucleic acid is determined by the ordering of its nucleobases along the sugar-phosphate backbone that constitutes the bonds of the nucleic acid polymer.  Yet primary structure tells us nothing about the larger, 3-dimensional shape of the molecule, which is vital for a complete understanding of nucleic acids.\n",
    "\n",
    "Watson and Crick proposed the following structure for DNA: \n",
    "1. The DNA molecule is made up of two strands, running in opposite directions.\n",
    "2. Each base bonds to a base in the opposite strand. Adenine always bonds with thymine, and cytosine always bonds with guanine; the complement of a base is the base to which it always bonds.\n",
    "3. The two strands are twisted together into a long spiral staircase structure called a double helix.\n",
    "\n",
    "Because they dictate how bases from different strands interact with each other, (1) and (2) above compose the secondary structure of DNA. (3) describes the 3-dimensional shape of the DNA molecule, or its tertiary structure.\n",
    "\n",
    "In light of Watson and Crick's model, the bonding of two complementary bases is called a base pair (bp). Therefore, the length of a DNA molecule will commonly be given in bp instead of nt. By complementarity, once we know the order of bases on one strand, we can immediately deduce the sequence of bases in the complementary strand. These bases will run in the opposite order to match the fact that the two strands of DNA run in opposite directions.\n",
    "\n",
    "\n",
    "The reverse complement of a DNA string is formed by reversing the string and taking the complement of each symbol.\n",
    "\n",
    "We must reverse the string in addition to taking complements because of the directionality of DNA: DNA replication and transcription occurs from the 3' end to the 5' end, and the 3' end of one strand is opposite from the 5' end of the complementary strand. Thus, if we were to simply take complements, then we would be reading the second strand in the wrong direction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5dc71244-290a-4d1a-83b1-66243b2604db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a DNA sequence: CCTCAATGTAGTCCTCGTTAGTGAACCATTCAACCAGACATTCGGACTACTGCGGTACCGTCAGTCTTCAGTTGAAGCAAAGGTACCTACGGGCCGGATTCACTTGTTGATAATACCGATAGTGCTCTGGAAACCAGGCCTAGCGTTCCCGAAGTATTCGTTAGGTCTAGCGGTAGACCCGGCAGGCCAGGAGGAACTGGAAAATTGTACGGTAAGCCAGCTGTATACCGCTAGGTTACCGTGCATCTAGCTACGAGACTATGTCAACATCGTCGTCGCCGATACACCTCATCGCAATTGTATCAGGGTAATTTACGGCATTAACGAGGAAAAAGTACCGAAATGGCGCGCGCCGGTACCAAGTGTTTGGACGGTCTATTCACTGACCGTAACGTGCGACGCTGTTGAGGCAAGGACGCGGCGAATATAATTCGGACGCGGAGGTTGAGAGTGCCATGTCAAACACCACAAGGACCAATCAACTTGGGCAGTCCCTCTGGGTATTAACGGAATGTGTAGATACTACGTTGCACAGTTTTAATCCAATGGTACCATGACGATAGAAAAAGCTAATTTCACACTCTGGGGACAGTGATGCGGTTTGCGCCTTCGCACGTAACATGACTGTAGCCACTTGGGCTTGCCAGAGATGCCTCATTGTTTAATCCTCGGAAACGCTCAAAGGGCAATTCTGACATCTCGATGTTCACTAAACGCGTCACCTTCAAATTAAGACCTGAATTAAACTGATGCCCGGATAGTGCGATTACGTCACCAGTTACACGATATGAGACGGAACCTTCCGTCCTTCTTGTCGCTGTGACCTGACGTACATGAGCTGGGATGAACTGCTTCGAAATGGAGGACCTCACTAG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTAGTGAGGTCCTCCATTTCGAAGCAGTTCATCCCAGCTCATGTACGTCAGGTCACAGCGACAAGAAGGACGGAAGGTTCCGTCTCATATCGTGTAACTGGTGACGTAATCGCACTATCCGGGCATCAGTTTAATTCAGGTCTTAATTTGAAGGTGACGCGTTTAGTGAACATCGAGATGTCAGAATTGCCCTTTGAGCGTTTCCGAGGATTAAACAATGAGGCATCTCTGGCAAGCCCAAGTGGCTACAGTCATGTTACGTGCGAAGGCGCAAACCGCATCACTGTCCCCAGAGTGTGAAATTAGCTTTTTCTATCGTCATGGTACCATTGGATTAAAACTGTGCAACGTAGTATCTACACATTCCGTTAATACCCAGAGGGACTGCCCAAGTTGATTGGTCCTTGTGGTGTTTGACATGGCACTCTCAACCTCCGCGTCCGAATTATATTCGCCGCGTCCTTGCCTCAACAGCGTCGCACGTTACGGTCAGTGAATAGACCGTCCAAACACTTGGTACCGGCGCGCGCCATTTCGGTACTTTTTCCTCGTTAATGCCGTAAATTACCCTGATACAATTGCGATGAGGTGTATCGGCGACGACGATGTTGACATAGTCTCGTAGCTAGATGCACGGTAACCTAGCGGTATACAGCTGGCTTACCGTACAATTTTCCAGTTCCTCCTGGCCTGCCGGGTCTACCGCTAGACCTAACGAATACTTCGGGAACGCTAGGCCTGGTTTCCAGAGCACTATCGGTATTATCAACAAGTGAATCCGGCCCGTAGGTACCTTTGCTTCAACTGAAGACTGACGGTACCGCAGTAGTCCGAATGTCTGGTTGAATGGTTCACTAACGAGGACTACATTGAGG\n"
     ]
    }
   ],
   "source": [
    "# sample dataset: AAAACCCGGT\n",
    "# sample output: ACCGGGTTTT\n",
    "\n",
    "# take a DNA string s of length at most 1000 bp\n",
    "# return the reverse complement sc of s\n",
    "\n",
    "##VARIABLES\n",
    "# input DNA sequence\n",
    "#s = 'AAAACCCGGT'\n",
    "# insert DNA sequence\n",
    "s = str(input('insert a DNA sequence:'))\n",
    "# create accumulator variable for reverse complement of a DNA string\n",
    "sc = ''\n",
    "\n",
    "##PROGRAMME\n",
    "# for each base in reversed DNA sequence string\n",
    "for base in s[::-1]:\n",
    "    # if base is 'A'\n",
    "    if base == 'A':\n",
    "        # sc becomes sc + 'T'\n",
    "        sc = sc + 'T'\n",
    "    # or if base is 'T'\n",
    "    elif base == 'T':\n",
    "        # sc becomes sc + 'A'\n",
    "        sc = sc + 'A'\n",
    "    # or if base is 'C'\n",
    "    elif base == 'C':\n",
    "        # sc becomes sc + 'G'\n",
    "        sc = sc + 'G'\n",
    "    # otherwise\n",
    "    else:\n",
    "        # sc becomes sc + 'C'\n",
    "        sc = sc + 'C'\n",
    "\n",
    "# print complementary DNA string\n",
    "print()\n",
    "print('the complementary DNA string, sc:', sc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c91ef029-c18b-40d1-8e76-f9752b64e239",
   "metadata": {},
   "source": [
    "## REVC - Complementing a Strand of DNA (Method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "025d1132-4ef8-4953-9cb5-b7372bb684bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a DNA sequence: AATAGGGCTGAGGCCTACTAGCATTGAATTTTTATGCATATCATCGCCTTACCGCTGAAGTCACAAGTAGATCAAGTGGAAATCTGAACACGTAACAAAAACATGGCCGTATTCTCGTTTGTTTCATACTGCGGAGGTATTCTATTCTGGCGTGGTTTGACATCGCGGTTGCCGGCGGACAAGTATGTCTTTACAGCCGTTTGCCTAGCCCCAGTGATAGAAGACGTGACCTCACCTTTAGCGCAGGTAACAACTTTTGAAAAAAGCTTCTATGGCATCGTATCGTCTTCCGAAAATACGAAGACGTTGGAAACGGATTCATTCGACAACGGCAACATGGATATTGATGAGTATTGACCATCTTGTAAATATAGAGCTCGTCATCTTGATGCATAGTCGTACCGGATGCACCGACGGCAAGGGGCCGTCCTTTCAATACGTTTAGTGCTGCGCAGAAGCCATTGTCTGCTATATGCTACCACCCTCATGCGGCCCAGCGGCATTGCCGACTTTGCTCTGCAAATCACTTGCCGGGAGCAGGTCGCCGAAACCAAGACAGTAGAGCCGGATAACTGTATATTCGAGCACCACTTCTCAATAGGAATGGTAAGCGTCGAATTCCCCAAGAAAAATGTGACTCAATGACTGGCGTGGATAACTAGGGCCTTGGCACCAGGCAAGCCTCCACGCTACGTTTTAGCGGAGGGATCCGCACAATATGGCCATAGGTAACCAGCGATTTGTCATCGGACCTTTCCTTAACGTGAACCGCTGGGTGAGGTAGTTCATCCACACTTTAATCAAACCGCTTGGATGAGCTGACTGGTTCCTGCAACTTTGGGCAGAGAGCACTCCTGATCTAGTATTTACGCTCTCCGGTGGCAGGCTGTCAGCTGATGGGGATCCAACAGCACGCTGTAAGAAGCGCGGGAGAGGGGCCGAGTATCAAATAATCTTCTATG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "the complementary DNA string, sc: CATAGAAGATTATTTGATACTCGGCCCCTCTCCCGCGCTTCTTACAGCGTGCTGTTGGATCCCCATCAGCTGACAGCCTGCCACCGGAGAGCGTAAATACTAGATCAGGAGTGCTCTCTGCCCAAAGTTGCAGGAACCAGTCAGCTCATCCAAGCGGTTTGATTAAAGTGTGGATGAACTACCTCACCCAGCGGTTCACGTTAAGGAAAGGTCCGATGACAAATCGCTGGTTACCTATGGCCATATTGTGCGGATCCCTCCGCTAAAACGTAGCGTGGAGGCTTGCCTGGTGCCAAGGCCCTAGTTATCCACGCCAGTCATTGAGTCACATTTTTCTTGGGGAATTCGACGCTTACCATTCCTATTGAGAAGTGGTGCTCGAATATACAGTTATCCGGCTCTACTGTCTTGGTTTCGGCGACCTGCTCCCGGCAAGTGATTTGCAGAGCAAAGTCGGCAATGCCGCTGGGCCGCATGAGGGTGGTAGCATATAGCAGACAATGGCTTCTGCGCAGCACTAAACGTATTGAAAGGACGGCCCCTTGCCGTCGGTGCATCCGGTACGACTATGCATCAAGATGACGAGCTCTATATTTACAAGATGGTCAATACTCATCAATATCCATGTTGCCGTTGTCGAATGAATCCGTTTCCAACGTCTTCGTATTTTCGGAAGACGATACGATGCCATAGAAGCTTTTTTCAAAAGTTGTTACCTGCGCTAAAGGTGAGGTCACGTCTTCTATCACTGGGGCTAGGCAAACGGCTGTAAAGACATACTTGTCCGCCGGCAACCGCGATGTCAAACCACGCCAGAATAGAATACCTCCGCAGTATGAAACAAACGAGAATACGGCCATGTTTTTGTTACGTGTTCAGATTTCCACTTGATCTACTTGTGACTTCAGCGGTAAGGCGATGATATGCATAAAAATTCAATGCTAGTAGGCCTCAGCCCTATT\n"
     ]
    }
   ],
   "source": [
    "# sample dataset: AAAACCCGGT\n",
    "# sample output: ACCGGGTTTT\n",
    "\n",
    "# take a DNA string s of length at most 1000 bp\n",
    "# return the reverse complement sc of s\n",
    "\n",
    "##VARIABLES\n",
    "# input DNA sequence\n",
    "#s = 'AAAACCCGGT'\n",
    "# insert DNA sequence\n",
    "s = str(input('insert a DNA sequence:'))\n",
    "# reverse the inserted DNA sequence\n",
    "rs = s[::-1]\n",
    "# transform into lower case letters\n",
    "rs = rs.lower()\n",
    "\n",
    "# in one go\n",
    "#s = str(input('insert a DNA sequence:'))[::-1].lower()\n",
    "\n",
    "##PROGRAMME\n",
    "# raplace each lowe case letter with an upper case letter of a complementary nucleobase\n",
    "sc = rs.replace('a', 'T') # firstly use rs and insert result into sc\n",
    "sc = sc.replace('t', 'A') # from this point onward work with sc only\n",
    "sc = sc.replace('c', 'G')\n",
    "sc = sc.replace('g', 'C')\n",
    "\n",
    "# print complementary DNA string\n",
    "print()\n",
    "print('the complementary DNA string, sc:', sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6135267b-d416-47cc-9573-093df353fc20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method descriptor replace:\n",
      "\n",
      "replace(self, old, new, /, count=-1) unbound builtins.str method\n",
      "    Return a copy with all occurrences of substring old replaced by new.\n",
      "\n",
      "      count\n",
      "        Maximum number of occurrences to replace.\n",
      "        -1 (the default value) means replace all occurrences.\n",
      "\n",
      "    If the optional argument count is given, only the first count occurrences are\n",
      "    replaced.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(str.replace)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a83862bb-072f-49d2-bb4d-a69cb458c0dd",
   "metadata": {},
   "source": [
    "## HAMM - Counting Point Mutations (Loop/Con)\n",
    "https://rosalind.info/problems/hamm/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a6a8c0-c6fd-4f05-9e3b-28dc71067947",
   "metadata": {},
   "source": [
    "### Theory\n",
    "\n",
    "A mutation is a mistake which occurs during the creation/copying of a nucleic acid, in particular DNA. Sum of micro-mutations over many generations might lead to positive macro effects.\n",
    "\n",
    "The simpliest and most common type of nucleic acid mutation is a point mutation, which replaces one base with another a a single nucleotide. In the case of DNA, a point mutation must change the complementary base accordingly.\n",
    "\n",
    "Two DNA strands taken from different organism or species genomes are homologous if they share a recent ancestor; thus, counting the number of bases at which homologous strands differ provides us with the minimum number of point mutations that could have occurred on the evolutionary path between the two strands. Minimising the number of point mutations separating two species needs to be done, since the principle of parsimony demands that evolutionary histories should be be as simply explained as possible.\n",
    "\n",
    "\n",
    "The Hamming distance between two strings having the same length is the minimum number of symbol substitutions required to transform one string into the other. If the strings are given by s1 and s2, then we write the Hamming distance between them as dH(s1,s2).\n",
    "\n",
    "We can compute the Hamming distance by visual inspection: the Hamming distance between two strings is simply the number of positions in the strings at which corresponding symbols differ. For example, the Hamming distance between the two strings in the figure below must be equal to 7."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "cfd93272-9346-44d3-b076-c950f532a3b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "first DNA sequence: TAATGTGCAGGTTACGTACATGGCTTACACTGGGCGAAAGACTGTTGCAATGCGTTCCTCCCTCGCAGACTCCTACTACGTGTACCTGGTAGCAGGGATCCGGGGAGGACTAGACGCCTGTCTCAATAACGGATAGGAGCGGCGCTATAATCGCGCGCATTTAGTTGAGAGACGCCCTCACCCGTACTAAACTCGGAAGTCCGGGATTCTAAGCCAGAGATTCGGGTGCTCCTACTTTATCGAATTGTGTAAGATCTAAAGATTGATTCGTCCGCGGAGGTCTCCCAGCCCAGTATTATAGCGTCAGTTTTTCACATGTCTGAGCCTAACAACGGCAAGCACGCCAGAAGAATACCTGCGTCCATCGACCGGAATAGCTACTGAACGTTGAGGTGACACGACCGGAAGACGTTTACCTCGGCTACTACATCTTTGACGAAAGACAATTGTAGCTTTCACGGTGCTTCTTACTATTCGAACGAATTCGTTCTAATACTTAAATTCAATTCACGTGTCGCCTTCAAGTAGTCACTAACTCGGCCTCTTGGTACTCCACGGTGTATTATAGGTGGCCTTTGACTCCTGGACGCTGGTTGGTCCAGCTCTGCGCTACTTGGTTCCCAGTTGAACGGCGCGGCGTCATACACGTCATCTAAGCGAGCTCGGCCGCACATTCCCGTTATGCGGACATCGGGTACTTAAATGCGCTTAGCCTGGGGTGTGTTTCAATGCAAGTAAGCTCCACGCCACAACGTTGACCCCCCAGGTACCTGAGACCGCAGCTTGTTGTCAGAGTATCATGAATCCGGCGCAAGAGGCCGCTACACAAGGCCGAAGGTCTTTGATGATCCTTCAATACTGTCCTCATTAAACCAACTTCGAAAAAGATTTCAGGGCATGAGCAACCCCTGAAAACCATATAATACGAACTGCTGCTGTGTGGTGACA\n",
      "second DNA sequence: ACCTGTGCGTGTGTAGTTCGTAGCGGAGACAGTGCCATGGATCATGGCAGCACGTTTCACTAGGTCAAAGTTCGATGTAGGCCACCTCCGGTGATCCATCACGGGCAGTCTACGCTCTTTTGTCAATACCATTGCAGTAGGACTCTTTCACATACCACTAATACTGGTGAGACGAGCCGGAGCGTAGTAGTTTCGCAGAGCCAGTTTAAAAATCACTGGATGCCGTAACTTCTACAATATAAGAAAGTCATAGCGTCGCACTACGATTATTTCGCGGTGATCTCAGAGCCTACTGCCAAGCTATTTGTTTTCAACGTGTGCCCACAGTTCTAATACAATGTTGAAAGACCGGGTAGTTGTCGCACCGGCTGGAACCCCCATTCAGAGGTGCGGCAACAAGACCGGAGCTCGGCTTCGCCGGTCTGTTCTTCTCCTCCCCCGGGGAAATGCCGCTTTCACGCTCCGTGGGCCTTTTGCTTCGTATTCTACTATACACCTACATGGAAGACATGCGCTGCCTAGAATCAACCAAGAGCTCGGGGGCGGGGTAATCGAAGGACTAGTATGGACGGCTCTAATCCCCTGGATGGCCGCTAGCCCCACGTTATTCCTAATTTACGCCCTTGTATAGTTGTGTCAACACAACCGGCCTCAGGCCTCCCGATCCGGGAGCAAACGGATACTAGGCGCCCAAGGACTAAATGGTCATTGGCCTGCGCTTTGCGTTCATGTTCTGGAACGCCGGCTCTTAACGATTATGCCGCAGGTTGCATCCAGCACGGTGTGATAACGGAATACTATGCATACGGTGTCATACAGTGCTACAGACGGAAGCAGGGCTTTGATCATACTCCTATTCAGTTTCACATAGCACAATATATACCGCAATGAATGTGAGAGAAATACGCCCCAGGAGCGCACGTGACGTATAGTTGGACTGCGTTGGCG\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Hamming distance is: 493\n"
     ]
    }
   ],
   "source": [
    "# return the hamming distance (number of mismatching nucleotide bases)\n",
    "\n",
    "##VARIABLE\n",
    "#s = 'GAGCCTACTAACGGGAT'\n",
    "s = str(input('first DNA sequence:'))\n",
    "#t = 'CATCGTAATGACGGCCT'\n",
    "t = str(input('second DNA sequence:'))\n",
    "d = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for particular position (index) in range of length of the string, do\n",
    "for position in range(len(s)): # alternatively a while loop\n",
    "    # if nucleotide base at the particular position in 's' string is not the same as the nucleotide base at the aprticular position in 't' string, then do\n",
    "    if s[position] != t[position]:\n",
    "        # add 1 to hamming distance\n",
    "        d = d + 1 # alternatively +=\n",
    "    #if not, then continue to with another iteration\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "# print Hamming distance\n",
    "print()\n",
    "print('Hamming distance is:', d)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6afb887a-98c4-4d15-857d-e37eb32375f0",
   "metadata": {},
   "source": [
    "## PROT - Translating RNA into Protein (Loop/Con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1e47575-a790-4acf-8d28-58a4bd1fd2c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein is a structure of ordered amino acids (polypeptide)\n",
    "# some proteins are composed of several subchains (polypeptides)\n",
    "# translatoin of mRNA -> amino acids -> protein\n",
    "# but, 4 RNA bases (A, T, G, U) -> 20 amino acids\n",
    "# thus, 3-nucleobase strings (codons) must be translated to form the amino acids (20 * 3 = 60)\n",
    "# possible codons: 4 to the power of 3 = 64\n",
    "# but, AUG is a starting codon, UAA, UAG and UGA are ending codons (do not code for an amino acid)\n",
    "\n",
    "# ribosome + tRNA -> peptides\n",
    "# tRNA = string of 3 RNA nucleotides on one end (anticodone) and amino acid on the other end\n",
    "# ribosome takes RNA molecule transcribed from DNA (mRNA), examines one codon at a time\n",
    "# but, at each step, the tRNA possessing the complementary anticodon bonds to the mRNA at this location, while the amino acid found on the opposite end of the tRNA is added to the growing peptide chain before the remaining part of the tRNA is ejected into the cell\n",
    "\n",
    "# 20 amino acids = 20 letters from the English alphabet, except B, J, O, U, X, Z\n",
    "# protein strings form from these 20 symbols\n",
    "# all possible combinations of A, T, G, U are needed\n",
    "# we need to add AUG at the start of the string and one of the 3 ending codons at the end (the \"stop\" codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "40c16d66-4185-46df-aea3-7e54b9268c6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAMAPRTEINSTRING\n"
     ]
    }
   ],
   "source": [
    "##VARIABLES\n",
    "# input RNA sequence as a string\n",
    "s = 'AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA'\n",
    "#s = str(input('insert RNA sequence:'))\n",
    "# 64 possible codon combinations from nucleotide bases (A, T, G, U)\n",
    "combinations = {\n",
    "    'UUU' : 'F',    'CUU' : 'L', 'AUU' : 'I', 'GUU' : 'V',\n",
    "    'UUC' : 'F',    'CUC' : 'L', 'AUC' : 'I', 'GUC' : 'V',\n",
    "    'UUA' : 'L',    'CUA' : 'L', 'AUA' : 'I', 'GUA' : 'V',\n",
    "    'UUG' : 'L',    'CUG' : 'L', 'AUG' : 'M', 'GUG' : 'V',\n",
    "    'UCU' : 'S',    'CCU' : 'P', 'ACU' : 'T', 'GCU' : 'A',\n",
    "    'UCC' : 'S',    'CCC' : 'P', 'ACC' : 'T', 'GCC' : 'A',\n",
    "    'UCA' : 'S',    'CCA' : 'P', 'ACA' : 'T', 'GCA' : 'A',\n",
    "    'UCG' : 'S',    'CCG' : 'P', 'ACG' : 'T', 'GCG' : 'A',\n",
    "    'UAU' : 'Y',    'CAU' : 'H', 'AAU' : 'N', 'GAU' : 'D',\n",
    "    'UAC' : 'Y',    'CAC' : 'H', 'AAC' : 'N', 'GAC' : 'D',\n",
    "    'UAA' : 'Stop', 'CAA' : 'Q', 'AAA' : 'K', 'GAA' : 'E',\n",
    "    'UAG' : 'Stop', 'CAG' : 'Q', 'AAG' : 'K', 'GAG' : 'E',\n",
    "    'UGU' : 'C',    'CGU' : 'R', 'AGU' : 'S', 'GGU' : 'G',\n",
    "    'UGC' : 'C',    'CGC' : 'R', 'AGC' : 'S', 'GGC' : 'G',\n",
    "    'UGA' : 'Stop', 'CGA' : 'R', 'AGA' : 'R', 'GGA' : 'G',\n",
    "    'UGG' : 'W',    'CGG' : 'R', 'AGG' : 'R', 'GGG' : 'G'\n",
    "} # see: https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables\n",
    "# accumulator string variable for codon extraction\n",
    "codon = ''\n",
    "# accumulator string variable for translation\n",
    "protein = ''\n",
    "\n",
    "##PROGRAMME\n",
    "# for position in range from 0 to number before the last number dividible by 3\n",
    "for position in range(0, len(s), 3):\n",
    "\n",
    "    # codon becomes codon with concatenated nucleotide base at position location\n",
    "    codon = codon + s[position]\n",
    "    # codon becomes codon with concatenated nucleotide base at position + 1 location\n",
    "    codon = codon + s[position+1]\n",
    "    # codon becomes codon with concatenated nucleotide base at position + 2 location\n",
    "    codon = codon + s[position+2]\n",
    "    # if value from combinations at codon location equals 'Stop'\n",
    "    if combinations[codon] == 'Stop':\n",
    "        # protein becomes protein with concatenating \"nothing\" (instead of 'Stop')\n",
    "        protein = protein + ''\n",
    "    # otherwise\n",
    "    else:\n",
    "        # protein becomes protein with concatenated value from combinations at codon location\n",
    "        protein = protein + combinations[codon]\n",
    "\n",
    "    # clear codon variable\n",
    "    codon = ''\n",
    "\n",
    "# print translated sequence, the sequence of the protein\n",
    "print(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52f85b96-8b02-4285-a5b7-7397f8c5920d",
   "metadata": {},
   "source": [
    "## GC - Computing GC Content (Loop/Con)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80eba586-d900-4443-a4f8-ea6d74c42484",
   "metadata": {},
   "source": [
    "### Theory\n",
    "A quick method used by early computer software to determine the language of a given piece of text was to analyze the frequency with which each letter appeared in the text. This strategy was used because each language tends to exhibit its own letter frequencies, and as long as the text under consideration is long enough, software will correctly recognize the language quickly and with a very low error rate. See Figure 1 for a table compiling English letter frequencies.\n",
    "\n",
    "You may ask: what in the world does this linguistic problem have to do with biology? Although two members of the same species will have different genomes, they still share the vast percentage of their DNA; notably, 99.9% of the 3.2 billion base pairs in a human genome are common to almost all humans (i.e., excluding people having major genetic defects). For this reason, biologists will speak of the human genome, meaning an average-case genome derived from a collection of individuals. Such an average case genome can be assembled for any species, a challenge that we will soon discuss.\n",
    "\n",
    "The biological analog of identifying unknown text arises when researchers encounter a molecule of DNA from an unknown species. Because of the base pairing relations of the two DNA strands, cytosine and guanine will always appear in equal amounts in a double-stranded DNA molecule. Thus, to analyze the symbol frequencies of DNA for comparison against a database, we compute the molecule's GC-content, or the percentage of its bases that are either cytosine or guanine.\n",
    "\n",
    "In practice, the GC-content of most eukaryotic genomes hovers around 50%. However, because genomes are so long, we may be able to distinguish species based on very small discrepancies in GC-content; furthermore, most prokaryotes have a GC-content significantly higher than 50%, so that GC-content can be used to quickly differentiate many prokaryotes and eukaryotes by using relatively small DNA samples.\n",
    "\n",
    "### Note on Absolute Error\n",
    "We say that a number x is within an absolute error of y to a correct solution if x is within y of the correct solution. For example, if an exact solution is 6.157892, then for x to be within an absolute error of 0.001, we must have that |x6.157892|<0.001, or 6.156892<x<6.158892.\n",
    "\n",
    "Error bounding is a vital practical tool because of the inherent round-off error in representing decimals in a computer, where only a finite number of decimal places are allotted to any number. After being compounded over a number of operations, this round-off error can become evident. As a result, rather than testing whether two numbers are equal with x=z, you may wish to simply verify that |xz| is very small.\n",
    "\n",
    "The mathematical field of numerical analysis is devoted to rigorously studying the nature of computational approximation.\n",
    "\n",
    "### Exercise\n",
    "The GC-content of a DNA string is given by the percentage of symbols in the string that are 'C' or 'G'. For example, the GC-content of \"AGCTATAG\" is 37.5%. Note that the reverse complement of any DNA string has the same GC-content.\n",
    "\n",
    "DNA strings must be labeled when they are consolidated into a database. A commonly used method of string labeling is called FASTA format. In this format, the string is introduced by a line that begins with '>', followed by some labeling information. Subsequent lines contain the string itself; the first line to begin with '>' indicates the label of the next string.\n",
    "\n",
    "In Rosalind's implementation, a string in FASTA format will be labeled by the ID \"Rosalind_xxxx\", where \"xxxx\" denotes a four-digit code between 0000 and 9999.\n",
    "\n",
    "Given: At most 10 DNA strings in FASTA format (of length at most 1 kbp each).\n",
    "\n",
    "Return: The ID of the string having the highest GC-content, followed by the GC-content of that string. Rosalind allows for a default error of 0.001 in all decimal answers unless otherwise stated; please see the note on absolute error below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "d901a996-16d7-40d5-b9be-737bc95ff4f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a file path with DNA strings in FASTA format: rosalind_gc.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_8505\n",
      "50.239234\n"
     ]
    }
   ],
   "source": [
    "##ABBREVIATIONS\n",
    "# gc = guanine and cytosine\n",
    "# nb = nucleotide base\n",
    "\n",
    "##VARIABLES\n",
    "# input DNA strings in FASTA format\n",
    "path = str(input('insert a file path with DNA strings in FASTA format:'))\n",
    "mode = 'r'\n",
    "\n",
    "fasta = []\n",
    "current_header = None\n",
    "current_seq = []\n",
    "\n",
    "# ChatGPT\n",
    "with open(path, mode) as file:\n",
    "    # looping through lines and removing \"\\n\"\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "\n",
    "        if line.startswith(\">\"):\n",
    "\n",
    "            # in the first iteration it is None, skipped; stores previous, if it is a header, it is added, if not sequence by sequence is stored until a new line appears that makes the whole sequence to be added and a new header to be stored\n",
    "            \n",
    "            if current_header is not None:\n",
    "                fasta.append(current_header)\n",
    "                fasta.append(\"\".join(current_seq))\n",
    "\n",
    "            # new record\n",
    "            current_header = line\n",
    "            current_seq = []\n",
    "\n",
    "        # build sequence\n",
    "        else:\n",
    "            current_seq.append(line)\n",
    "\n",
    "# adds the last sequence, because a sequence is added only when there is a new header\n",
    "if current_header is not None:\n",
    "    fasta.append(current_header)\n",
    "    fasta.append(\"\".join(current_seq))\n",
    "\n",
    "#fasta = ['>Rosalind_3510','ATCACAGGATACAAACGAGGACTGCGGCGGCATCAAAATGGATCGGGAGCCCCAGTCTACGCCTATTGTTGTTCGGCCGGATGTGCTTGTCGGGCGGCAGTCCGTTGTAGGGGTGCTTTACAGCGAACTGAATATAGCACCGAAAGAGCCAGGAGTGGCATGCCAGTATAATTATGTTCGCTAATGTGCTTGGTTTATGTTTTCATGACGCAGGGTATGTAAAAGGTGTCCCCTGCCCGCACTTATGGGGCCCGTATTCATGAGAGAGGGATTTGCCAAAACAGGAGCACCGAACAGGAGTGAGAAATAGACTGGAAGCTGGACTCGTGACGACGTAGTAGACGGGGCCCAGAGTAGCACTTTTGAAGCGGAGCCTAGGCCTGCGGGCAGCACCGCCATCACGCACCCCGTATACACAGTCCTGTGGGCCGTGACGACGCACGTCGTTACTAATGCATTTGCTTCGGGGCACTACCCGTCAGCATGGCTTTTTGCCTCTGAAAGAATGTGTCTTATAAACTAATAAGACAGGAGGGTACCACGTATCCACCCCTCTATACAGAAGACGTGGGAGCGATTGTGTTGTACCGGTAACCGTCTTACCACGAGCCCACTGTCGGAGTACATGCGCATCACGGGCACGTAACTACCATGCGGGGGTCTAACCTACGATCTTAGTTTTTTGATGCAAGGGCGCTAAACCATCCCCCCGCTGACGGTGTTGACTACGTTTTGCACAGGCTCCTTGTAACAAGACACGAACTACCAGACCAAGTAAAAAAGTGTATACTGTTGCTGTTAAACCATGAGGCAGAATAAGGGCTGTTGTCGATGCTTGCCAGCGGTCCCCCGTTTAGGTGTTGTATTGTTCTACGTTGGCTTTACAGCCACCTGTCCGCAACCTTACCGTCGCTTACTACGGATCCTCGTAAGAATATAGTCCTCATGCTAAACACGACTTATCCTTCGACCTCCAGGAG','>Rosalind_0901','GTTGGTACTGATTATGTCGCTCCTCACAATAGGTCACCGGGCGTACCATGAATCTAGAGTAAGCATGCCGGAGGGATAGGCAGGCTTGTTAACTCAACAAAGCAATAAAACAGACACCTGCCAATAATCCATGCAAACCGAAGTTTACGCATGTCCACGCATCCCGGGAGAGGATTTTAAGAGGGTACCTCTGCGGAAAACTTTGGAATTCAGAGTAAAACTGTTTAAGCACGTAACCGCCTACATTAGTTTTTGGTAGTAACGACCATGTCAAGATTAGGTGACCCTCAGAGTGGCTGTTCACCTTTGTCCTCGAAGATGTACCAACAAGCTAGTCGGGGCCTGGTTTGCACACGGTTTACGGAATAAGCGCACTCTTCTAGGCGACAACAGTCCCCTTGTAAAGGCCAGGTTAACCTGTGGTGTCACCGTAAGTATTTAACAACGGCTGATCGGGTCGGGTGCACGTGGAAGAAGCGGCTTTCAACTCTCTCAGTAAGACGAATGCCTGGCTAAAGAGTTGGAGAGGTGCTAAAAACGGGAGTTATACGTGACGCGCCAATGGAAGCCTCCCCGATGAGCGGGCCCTCCGAGACACTCCTACCCAAAGACCAAGTCGACGGTCAACACCAGAGCGGCTGACGGACGACGTATTCGACCGATACTTGCGAGGCTCTGGACAACGATACTCTGCGTCCGGGCAGTTTTATCTAGATCAAGAGGCAACAGGCAAACTTCAACTATACACTGGCGCGTTAGAGAGTTCTCTTTTTTATACTGGTTTGCAATTTGACGTTGTAATGTAGACTGCGTGGTCGCCATTCACATTCCGCGCGGTGAGCCCGGCCTAAAGCTCGGCAGCCTACACTAACCCAAAAGCAACATCTTGACTGTATGTACGTCGCCGGTAATCTGGCAATCCTAACTAACTACCCTTGGAAAAAATCCATCGTCAGTGTCCCCATGTTCACCAGTTTTTATTTTTACGGGTGG','>Rosalind_3533','GGCATCCCTGCGTAGTTAAAGACGCATTGGGGTGATACCGGCGAAATTATGACGCTGAGTGTCTTCGCACTTGAGGATCGGATAATTTAATGCAAGTTTAAGTCGATACGTGTAGGAACAGGCCTGGCGGCAGTCTTGCGGAAACTTCAAGCTAATCAGGTTAACATTTAGCAAGTGCGGCCGACGCCGGAAGCCCGTTTCATTCATTCCTGTGTGTTCAACGTCCTTACGTTGAGTTGGAGGGGTTACTCGTATCAGTAGTAATTACTGAAATTCACATATGGTCTAGACCACTGAGAGGTGACACAAAGTCCTGATCTATAACTTCCTACACCTATTAGCGGGTCTATGGTAAGGCTGGGATCATCAGTTTCGTCACGCGCTCCATTTATACGAGGTGTCTTGAGCATGTCGGTAGAGGGTGGCTAGCGCACGAACGAAAGAGCTTTCACCTTTACGTCACCGCGGTTACAGTCCTAGGCACTTTCGCAGTTCGCTTTTTCGTCGGAACTAGGGCTTCTTTCTCAGCTACTCGAGCAACTGATGACTATCGTCGGGGAAAGCCCTCTCAGACGTATGCACTTCAAACGCCTGGATGGGCAGGTGAAGTGATTTGAAAATTTTCGTTAAAATTCCATGATGTGACTACGGCGACGATGTACTGTGATTATCCCCAATTTTGGGATATCTAAGGCGCTCCGAACCCTTGCCGGACCTGTACCTAAAAGAAGCAAATCATAATATCTTTAGGTTCACGGGTGGGAAGTGCTTTTCCAGTAAACTTGCAGAATGTGCTAGTTTATCAACGACGATACACGAGACCCGATCGGGTATGGAGACCTTCTGGAGCGATTAGCCGGAGCCTTAGGAGCCCATTTTTAATGCTCTATAAGTTTCGGGTCTCTAGTGGTCGGTTGGTGTCGCTCAAG','>Rosalind_8710','CTTTACTCAAGTCCATCATGTGCACAATTATTGACTTCATGTCCCGTCAGCGGCGTTCCAGGGACCCTGTGCGCAAGATCAATTTCTTACACTTAACGGAAAGATAGTCACTGGCCTTGTCACGGAGACCCTCGTACAGTGAAATCTACTCTCTATCGTGATAGGCTAGTAATCTAGGACAGTCGGTATCATACGAGTCCACGGTAACGATGGTTTGATATGGCTACTTGTAACGTAGGCCTAATCGATTCTGTAGCAAACAATCTGTCCCACGACTACGGGGCCAGCCGAGTATTCAAGGATTCCTTGCCGATTTCATGCCGCATTGACGAACACATAGCGAGGGGACACGTGGTCAGCTGAGAGTCATCACCACTAAATGCGCCAGGATACTAGCCGGTGGAGCGGTCCTCGTTCTCGTTCTAGTGCTTTACAGGATTCGGACCGGTTGTACTTATGAGTACAGAAGCGCGCAAGATAGGAGGGCAGGTACTACTATCAACCCCACGTTAAAATAGACGTGCCAACGCCCCATTTTGGCGGGCCCATCATTTCTTCGTTACCACCATCTCGTACACGGGGTCGTGTGCAACAATGTCAAATCGTACCACCTTAATCAGGTAATTTTGGCCGACTTCCATCGGACGGCGGTCTCACATTTGGCTTCGATCGCATATTCGCCAAGATCTTCCCGAGAGCTTGTTGGTCAGCCGCGCGACCCACACCTGCGGATCATTCGCCCCTGGGCTAGTTGGTGTGATGTCGGAGCCGGAGTTGAGCCCCGCGGCACAGCGGATGTAGGAGGGCCAGAGCCCCGTTTACGAACGTCCAGTTTTGACTCTCCCGTATTGCTTGCTACCCAAACCGTATATATTTCGTGGTCTTCCGATTGGTGCGGATCTGATAGCTAGACCAACCGCCCTTCGGCAATAAATATCG','>Rosalind_1706','CGATAAACAGTTAACTCGGCAAGGTGAGAAAGCGCAAACTGTCTTGAACCAAATAGGAGCTATGTCTGATGAAATTGATTAAACATGTTTATGTCCTATCCATGGCCCACACAGAGGCAGATAGGATCAAGCGATCCTGACCAGGGGGGAACAAGTTCTAGAAAAGGAGAGGACCTTGTTTAGAAAGTTTAAAACCTAAATACCGATCCTGTCATTGCCTTTAAGGCATCAGCATTATAAGCCAGTAGTACGGAGTTCCTGCGAATGGAGTTGTAACAGAGTTGGTAGATAGTTCTATTTAGGGTGCGCTCGTCTGGCTGTCGATGCACATAATAAGGGCGTCATAATCCGCACCGACGCGTATCCGCTTTTGGGTGATGCAGAGGCTGTACCCATTTGTACTCGACCCATGCGAAAACGAGGGCTTGTCGATAAATTTTCTTAGTTTTACTCAAGAGTTATTAGATAGCCCCGCCTTTTATCTTTAGTGCTACAGAAAAAGGAGGTCCAGGAAGCATCTCTCTCGCTGTTCTGTAATGGTGGCCGTATCCCTACTCGTGTCGCATTAGGCACAGGCTCTACTGGAAACGATGCGGATGCACCGTAAGCAACAGAACCGTAACGAACAGCACCCCGCTACCTTTTATCACTACACGTCCAGAAGGCCCAAGGCGACCTGAACCTTTAATTGGAGATTCATTGATCTCTGAACTTGCCACGTGAATGCTGATTTTTTTCCCTTCACATCTGAATAACCATCCTATGGCTTCTAGCGAGTGTGAGGAGTGACGAAGACTGGTAAAGAGGTATCAGCCATTCTTTCTACCTGCCGAGATTTCGCTTGGACCGTCAGGCGGAGCCTCGAA','>Rosalind_9457','TAGAATGACTGCGTCCGATATCTTACAATTTAAGTAAAACTTGTCAACCACCAGCTGGAGACAGACAACTATCACGAGCGCAGCAGCCGGGAATGGTCTTAGAATCTTGAACCACGCTGAAAGTCGGACTTAAGTCTGCGGTGGCGTGCAAGAAGATCGCACCAGGTGTGCCACTACCTCTTGTAAGGGGCTTAATGTGCCACCTCTCTCTTACAATTGTGCGGCCGGCCAGTTATTATCGGTAATACATCTTAGTATTTGACGGAAGTGATGAGCCCAATCATTCATCCGGCGCAGACCTCTCCATATACCACGCATATCAGAGTAGGCTTTTTGGGCGAGAACTTTCCGTACACTTAATTCGAGTGATCAGATCTTCGATGCTTAGTGAACTAGTGCAGACAGAATTCCAATAAGACGGCTGCCGGAAGTATCTTCGCCAGATGGCGTTGACCTTGGAACCGGCTGAAATCAGTTTGAACTGGTCAGGCACTACTATAGCAGATCCATTTTAGGCATAAGGCTGTCATGCTGAAAGGTTGACAGGGCTCTCATTCGCCCGCAACTAACAGAAGTCGAGCCATGACCATCCCCGTCGTTTAGCTTCCGCACCCATTCCGACCATCATGGGTACGACCGAAGTCTCACCAAAACGACCGGGACTATTTATAATCTAGAACAAAGCACCACTTTGGTCCGCATGCAGGGTCACAACTGCAATGTAGGATGATACAGTTAAAAGATCGATGCACATATTAAGAACGAAATACCCTACGTCGCACACCCCACCTAGCACTGGAAGAGACTCATCACAGGCCATTGTACCCTTAGAGTAATTGCCTTGAATAG','>Rosalind_4585','CCTTTCCGCAAGGCCAACAAGGTACCAAAATTGAAGTTTGGCCAGCTAGCAATAATACTGAATTGATATTCGCGAAACTCGCTTGAAGTTATATGAGCGATCAAATCATAACTCAGTATCCCTAATAAGGAGCGCTGCGCTCAGAGAGTGTTACGCCATGTGAACGGCCTGACGAGGGAATGTTCCGTACGACAAAGATGCTCCCAACGCCTTAGGTCCCACCGTAGAACCTGACATGGCATACATGCCCAGCGCGCTGGACCGCTCACTCGGTCTACTGGGTACGTGACTTAGGGTTCGTGAGATTATCGCATTAGTCTTAACTAGCGTTCGATAACTGGTAAATTAGAGCGGTCGGTAGAGGCATATCAAGCTGGTTTGTGACACACGGACATGGAACGACTCTAGCTTTTGTTGTGATATGAGGTGGTGCTAACGGACGTGTGCGGGCGCTCACAACCCGCAGAACACCGACGCCAGAGTTCCATTTTTCCCTGGAGCCGCATACTGTCCACTGTATTATAGGAGAAATTTATTCTATGCCTGTACTATGCCATCTCTCACGTTCCCAGACTTGACCCTTATAAACATGTATATGATTTGCATAACTTTTCCGGTTTCCATCATGCTTAACTACTCAGAAGACTGAACTGGAGATGGCTTCGATTGCGTATACGCCGTGACACAGCCCCTGAAGGACCACGATCACGCAGCATAGTGTACTGCCGGTTACCCATCATAACGTGTCCAGCAAACCGCCAGTATTGTAAAATCGCATAGCTAGGCCATGCCATGTAAATCTTGAGTGCCAGATCCCGGGCCATAGCGGCATAGTTAAACATTTATGCCCATTACTCCCTTTTCAGAACAAAAATCGACCTATCGTCTGGAATCCCACCACACAATGCTGAGATTTCACACCAACTCCAACAGTAGGATTCCAGGGGCCGATACCCGCAC']\n",
    "\n",
    "#fasta = ['>Rosalind_6404', 'CCTGCGGAAGATCGGCACTAGAATAGCCAGAACCGTTTCTCTGAGGCTTCCGGCCTTCCCTCCCACTAATAATTCTGAGG', '>Rosalind_5959', 'CCATCGGTAGCGCATCCTTAGTCCAATTAAGTCCCTATCCAGGCGCTCCGCCGAAGGTCTATATCCATTTGTCAGCAGACACGC', '>Rosalind_0808', 'CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGACTGGGAACCTGCGGGCAGTAGGTGGAAT']\n",
    "# accumulator variable for \"GC\" count first, after counting, for content in percentage\n",
    "gc_content = {}\n",
    "# accumulator variable for nucleotide base count\n",
    "nb_total = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for each string (only) in a range from the first one to the last one\n",
    "for string in range(1, len(fasta), 2):\n",
    "    # create a dictionary key (id)  and value (0)\n",
    "    gc_content[fasta[string-1][1:len(fasta[string-1])]] = 0\n",
    "    # reset to 0\n",
    "    nb_total = 0\n",
    "\n",
    "    # for each nucleotide base in the current string\n",
    "    for nb in fasta[string]:\n",
    "        # add 1 to total count of nucleotide bases\n",
    "        nb_total += 1\n",
    "\n",
    "        # if nucleotide base is \"G\" or \"C\"\n",
    "        if nb == 'G' or nb == 'C':\n",
    "            # add 1 to the value of the corresponding key in the dictionary\n",
    "            gc_content[fasta[string-1][1:len(fasta[string-1])]] += 1\n",
    "\n",
    "    # calculate percentage ((gc count / total number of nucleotide bases) times 100)\n",
    "    gc_percentage = (gc_content[fasta[string-1][1:len(fasta[string-1])]] / nb_total) * 100\n",
    "    # replace gc count with the calculated percentage\n",
    "    gc_content[fasta[string-1][1:len(fasta[string-1])]] = gc_percentage\n",
    "\n",
    "# derive the highest percentage\n",
    "gc_highest = max(gc_content.values())\n",
    "\n",
    "# for each key and value in the dictionary (id : percentage)\n",
    "for key, value in gc_content.items():\n",
    "    # if value equals the highest percentage\n",
    "    if value == gc_highest:\n",
    "        # print key (id)\n",
    "        print(key)\n",
    "        # print value (percentage) with 6 decimal places (no rounding and keeping last 0)\n",
    "        print(str(value)[:str(value).find('.')+7])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1037f8a5-dc6f-4b3f-9677-24010bcd2f29",
   "metadata": {},
   "source": [
    "## IPRB - Mendel's First Law"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f4640c-5465-4c8f-810e-3b57fc7ff6d9",
   "metadata": {},
   "source": [
    "### Theory\n",
    "Modern laws of inheritance were first described by Gregor Mendel (an Augustinian Friar) in 1865. The contemporary hereditary model, called blending inheritance, stated that an organism must exhibit a blend of its parent's traits. This rule is obviously violated both empirically (consider the huge number of people who are taller than both their parents) and statistically (over time, blended traits would simply blend into the average, severely limiting variation).\n",
    "\n",
    "Mendel, working with thousands of pea plants, believed that rather than viewing traits as continuous processes, they should instead be divided into discrete building blocks called factors. Furthermore, he proposed that every factor possesses distinct forms, called alleles.\n",
    "\n",
    "In what has come to be known as his first law (also known as the law of segregation), Mendel stated that every organism possesses a pair of alleles for a given factor. If an individual's two alleles for a given factor are the same, then it is homozygous for the factor; if the alleles differ, then the individual is heterozygous. The first law concludes that for any factor, an organism randomly passes one of its two alleles to each offspring, so that an individual receives one allele from each parent.\n",
    "\n",
    "Mendel also believed that any factor corresponds to only two possible alleles, the dominant and recessive alleles. An organism only needs to possess one copy of the dominant allele to display the trait represented by the dominant allele. In other words, the only way that an organism can display a trait encoded by a recessive allele is if the individual is homozygous recessive for that factor.\n",
    "\n",
    "We may encode the dominant allele of a factor by a capital letter (e.g., A\n",
    ") and the recessive allele by a lower case letter (e.g., a\n",
    "). Because a heterozygous organism can possess a recessive allele without displaying the recessive form of the trait, we henceforth define an organism's genotype to be its precise genetic makeup and its phenotype as the physical manifestation of its underlying traits.\n",
    "\n",
    "The different possibilities describing an individual's inheritance of two alleles from its parents can be represented by a Punnett square.\n",
    "\n",
    "\n",
    "### Random variables\n",
    "A variable that becomes any value from a range of specified values. \n",
    "\n",
    "The probability could be (a frequency of something / total number of possible outcomes). For example (Pr(x)), a sac with blue and red balls in a sack, blue = 3/5 probabily and red = 2/5 probability. They some up to 1. Another example(P(y)), 1st is taken out, probability of the second one is determined from the remaining and if we wanted to know the exact probability, we would have to multiply their probabilites that lead to our wanted one and summed with another option (2 colours, one per each).\n",
    "\n",
    "\n",
    "### Exercise\n",
    "Given: Three positive integers k, m, and n, representing a population containing k+m+n organisms: k individuals are homozygous dominant for a factor, m are heterozygous, and n are homozygous recessive.\n",
    "\n",
    "Return: The probability that two randomly selected mating organisms will produce an individual possessing a dominant allele (and thus displaying the dominant phenotype). Assume that any two organisms can mate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "id": "51e92cb7-f61b-4364-ad17-c404161591a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.78333\n"
     ]
    }
   ],
   "source": [
    "##EXERCISE\n",
    "# Given: Three positive integers k, m, and n, representing a population containing k+m+n organisms: k individuals are homozygous dominant for a factor, m are heterozygous, and n are homozygous recessive.\n",
    "\n",
    "# Return: The probability that two randomly selected mating organisms will produce an individual possessing a dominant allele (and thus displaying the dominant phenotype). Assume that any two organisms can mate.\n",
    "\n",
    "##EXAMPLE\n",
    "# Input: 2, 2, 2\n",
    "# Output: 0.78333\n",
    "\n",
    "\n",
    "\n",
    "##VARIABLES\n",
    "k = 2\n",
    "#k = int(input('insert the number of homozygous individuals:'))\n",
    "m = 2\n",
    "#m = int(input('insert the number of heterozygous individuals:'))\n",
    "n = 2\n",
    "#n = int(input('insert the number of homozygous recessive individuals:'))\n",
    "\n",
    "##PROGRAMME\n",
    "def prob (k, m, n):\n",
    "    '''Calculates probability of getting an offspring with a dominant allele. Any option can cross with any option\n",
    "    k = integer of homozygous (AA)\n",
    "    m = integer of heterozygous (Aa)\n",
    "    n = integer of homozygous recessive (aa)\n",
    "    N is calculated within the function and corresponds to the total number of inidivduals.\n",
    "    '''\n",
    "    N = k + m + n\n",
    "    res = k/N + (m/(N*(N-1))) * ((4*k + 3*m + 2*n -3)/4) + (n/(N*(N-1)) * ((2*k + m)/2))\n",
    "    return round(res, 5)\n",
    "\n",
    "print(prob(k, m, n))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25bbeeb9-35b9-4b28-a188-7a83dd03b8c9",
   "metadata": {},
   "source": [
    "## SUBS - Finding a Motif in DNA (Loop/Con)\n",
    "\n",
    "### Exercise\n",
    "Given two strings s and t, t is a substring of s if t is contained as a contiguous collection of symbols in s (as a result, t must be no longer than s).\n",
    "\n",
    "The position of a symbol in a string is the total number of symbols found to its left, including itself (e.g., the positions of all occurrences of 'U' in \"AUGCUUCAGAAAGGUCUUACG\" are 2, 5, 6, 15, 17, and 18). The symbol at position i of s is denoted by s[i].\n",
    "\n",
    "A substring of s can be represented as s[j:k], where j and k represent the starting and ending positions of the substring in s; for example, if s = \"AUGCUUCAGAAAGGUCUUACG\", then s[2:5] = \"UGCU\".\n",
    "\n",
    "The location of a substring s[j:k] is its beginning position j; note that t will have multiple locations in s if it occurs more than once as a substring of s (see the Sample below).\n",
    "\n",
    "Given: Two DNA strings s and t (each of length at most 1 kbp). E.g., GATATATGCATATACTT\n",
    "ATAT\n",
    "\n",
    "Return: All locations of t as a substring of s. E.g. 2 4 10\n",
    "\n",
    "### Note\n",
    "Different programming languages use different notations for positions of symbols in strings. Above, we use 1-based numbering, as opposed to 0-based numbering, which is used in Python. For s = \"AUGCUUCAGAAAGGUCUUACG\", 1-based numbering would state that s[1] = 'A' is the first symbol of the string, whereas this symbol is represented by s[0] in 0-based numbering. The idea of 0 based numbering propagates to substring indexing, so that s[2:5] becomes \"GCUU\" instead of \"UGCU\".\n",
    "\n",
    "Note that in some programming languages, such as Python, s[j:k] returns only fragment from index j up to but not including index k, so that s[2:5] actually becomes \"UGC\", not \"UGCU\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "id": "4f58967a-1710-4a4b-9a7d-891bc6294230",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a DNA sequence: a\n",
      "insert a sequence to be found within the DNA sequence; must be no longer than t: a\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the string t is too long, please insert a shorter string\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert a sequence to be found within the DNA sequence; must be no longer than the string t: a\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the string t is located within the string s at the following starting positions: 1 \n"
     ]
    }
   ],
   "source": [
    "##ABBREVIATIONS\n",
    "# nb = nucleotide base\n",
    "\n",
    "##VARIABLES\n",
    "# input DNA sequence as a string \n",
    "s = str(input('insert a DNA sequence:'))\n",
    "#s = 'GATATATGCATATACTT'\n",
    "# input susbtring of s, not longer than t, as a string\n",
    "t = str(input('insert a sequence to be found within the DNA sequence; must be no longer than t:'))\n",
    "#t = 'ATAT'\n",
    "# accumulator variable which stores nucleotide bases\n",
    "check = ''\n",
    "# accumulator variable which stores real positions at which the string t was found\n",
    "position = ''\n",
    "# correction variable which acounts for the length of the string t and the real position where it was found accordingly\n",
    "correction = len(t)-2\n",
    "\n",
    "##PROGRAMME\n",
    "# check if the length of the string t is not more than the length of the string s\n",
    "if len(t) >= len(s):\n",
    "    print('the string t is too long, please insert a shorter string')\n",
    "    t = str(input('insert a sequence to be found within the DNA sequence; must be no longer than the string t:'))\n",
    "\n",
    "# for each nucleotide base in the length of the string s\n",
    "for nb in range(len(s)):\n",
    "    # check becomes check with concatenated nucleotide base\n",
    "    check = check + s[nb]\n",
    "    #print(check)\n",
    "    # if the string t is within the string check\n",
    "    if t in check:\n",
    "        # the string position becomes position with concatenated starting position and one space\n",
    "        position = position + str(nb-correction) + ' '\n",
    "\n",
    "    # if the length of the string check is equal to the length of the string t\n",
    "    if len(check) == len(t):\n",
    "        # the string check becomes check with 4 nbs without lastly added first one\n",
    "        check = check[1:len(t)]\n",
    "\n",
    "# if the length of the string position is 0\n",
    "if len(position) == 0:\n",
    "    # print message\n",
    "    print('the string t is not present within the string s')\n",
    "# otherwise:\n",
    "else:\n",
    "    # print message with positions\n",
    "    print('the string t is located within the string s at the following starting positions:', position)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df783fe1-2c1f-4616-b10a-58c6d2e1b465",
   "metadata": {},
   "source": [
    "## PRTM - Calculating Protein Mass (Loop/Con)\n",
    "\n",
    "### Theory\n",
    "In Translating RNA into Protein, we examined the translation of RNA into an amino acid chain for the construction of a protein. When two amino acids link together, they form a peptide bond, which releases a molecule of water; see Figure 1. Thus, after a series of amino acids have been linked together into a polypeptide, every pair of adjacent amino acids has lost one molecule of water, meaning that a polypeptide containing n amino acids has had n1 water molecules removed.\n",
    "\n",
    "More generally, a residue is a molecule from which a water molecule has been removed; every amino acid in a protein are residues except the leftmost and the rightmost ones. These outermost amino acids are special in that one has an \"unstarted\" peptide bond, and the other has an \"unfinished\" peptide bond. Between them, the two molecules have a single \"extra\" molecule of water (see the atoms marked in blue in Figure 2). Thus, the mass of a protein is the sum of masses of all its residues plus the mass of a single water molecule.\n",
    "\n",
    "There are two standard ways of computing the mass of a residue by summing the masses of its individual atoms. Its monoisotopic mass is computed by using the principal (most abundant) isotope of each atom in the amino acid, whereas its average mass is taken by taking the average mass of each atom in the molecule (over all naturally appearing isotopes).\n",
    "\n",
    "Many applications in proteomics rely on mass spectrometry, an analytical chemical technique used to determine the mass, elemental composition, and structure of molecules. In mass spectrometry, monoisotopic mass is used more often than average mass, and so all amino acid masses are assumed to be monoisotopic unless otherwise stated.\n",
    "\n",
    "The standard unit used in mass spectrometry for measuring mass is the atomic mass unit, which is also called the dalton (Da) and is defined as one twelfth of the mass of a neutral atom of carbon-12. The mass of a protein is the sum of the monoisotopic masses of its amino acid residues plus the mass of a single water molecule (whose monoisotopic mass is 18.01056 Da).\n",
    "\n",
    "In the following several problems on applications of mass spectrometry, we avoid the complication of having to distinguish between residues and non-residues by only considering peptides excised from the middle of the protein. This is a relatively safe assumption because in practice, peptide analysis is often performed in tandem mass spectrometry. In this special class of mass spectrometry, a protein is first divided into peptides, which are then broken into ions for mass analysis.\n",
    "\n",
    "### Exercise\n",
    "In a weighted alphabet, every symbol is assigned a positive real number called a weight. A string formed from a weighted alphabet is called a weighted string, and its weight is equal to the sum of the weights of its symbols.\n",
    "\n",
    "The standard weight assigned to each member of the 20-symbol amino acid alphabet is the monoisotopic mass of the corresponding amino acid.\n",
    "\n",
    "Given: A protein string P of length at most 1000 aa. E.g., SKADYEK\n",
    "\n",
    "Return: The total weight of P. Consult the monoisotopic mass table. E.g., 821.392"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "ba13f2d8-8a20-47b6-ba60-0c72de37688f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "821.392\n"
     ]
    }
   ],
   "source": [
    "##VARIABLES\n",
    "# input string\n",
    "p = 'SKADYEK'\n",
    "#p = str(input('insert a protein string:'))\n",
    "# wighted alphabet (w_a)\n",
    "w_a = {\n",
    "'A' : 71.037110,\n",
    "'C' : 103.00919,\n",
    "'D' : 115.02694,\n",
    "'E' : 129.04259,\n",
    "'F' : 147.06841,\n",
    "'G' : 57.021460,\n",
    "'H' : 137.05891,\n",
    "'I' : 113.08406,\n",
    "'K' : 128.09496,\n",
    "'L' : 113.08406,\n",
    "'M' : 131.04049,\n",
    "'N' : 114.04293,\n",
    "'P' : 97.052760,\n",
    "'Q' : 128.05858,\n",
    "'R' : 156.10111,\n",
    "'S' : 87.032030,\n",
    "'T' : 101.04768,\n",
    "'V' : 99.068410,\n",
    "'W' : 186.07931,\n",
    "'Y' : 163.06333,\n",
    "}\n",
    "# accumulator variable storing the sum\n",
    "p_weight = 0\n",
    "\n",
    "##PROGRAMME\n",
    "# for letter in p string\n",
    "for letter in p:\n",
    "    # p_weight becomes p_weight with added value from the dictionary at the letter position\n",
    "    #p_weight = p_weight + w_a[str(letter)]\n",
    "    p_weight += w_a[str(letter)]\n",
    "\n",
    "# print result protein weight of the string rounded to 3 decimal numbers\n",
    "print(round(p_weight, 3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfb06e09-1346-42fe-aa7c-6ab3175fa225",
   "metadata": {},
   "source": [
    "## FIB - Rabbits and Recurrence Relations (Loop/Con)\n",
    "### Theory\n",
    "In 1202, Leonardo of Pisa (commonly known as Fibonacci) considered a mathematical exercise regarding the reproduction of a population of rabbits. He made the following simplifying assumptions about the population:\n",
    "\n",
    "The population begins in the first month with a pair of newborn rabbits.\n",
    "Rabbits reach reproductive age after one month.\n",
    "In any given month, every rabbit of reproductive age mates with another rabbit of reproductive age.\n",
    "Exactly one month after two rabbits mate, they produce one male and one female rabbit.\n",
    "Rabbits never die or stop reproducing.\n",
    "Fibonacci's exercise was to calculate how many pairs of rabbits would remain in one year. We can see that in the second month, the first pair of rabbits reach reproductive age and mate. In the third month, another pair of rabbits is born, and we have two rabbit pairs; our first pair of rabbits mates again. In the fourth month, another pair of rabbits is born to the original pair, while the second pair reach maturity and mate (with three total pairs). The dynamics of the rabbit population are illustrated in Figure 1. After a year, the rabbit population boasts 144 pairs.\n",
    "\n",
    "Although Fibonacci's assumption of the rabbits' immortality may seem a bit farfetched, his model was not unrealistic for reproduction in a predator-free environment: European rabbits were introduced to Australia in the mid 19th Century, a place with no real indigenous predators for them. Within 50 years, the rabbits had already eradicated many plant species across the continent, leading to irreversible changes in the Australian ecosystem and turning much of its grasslands into eroded, practically uninhabitable parts of the modern Outback (see Figure 2). In this problem, we will use the simple idea of counting rabbits to introduce a new computational topic, which involves building up large solutions from smaller ones.\n",
    "\n",
    "### Exercise\n",
    "A sequence is an ordered collection of objects (usually numbers), which are allowed to repeat. Sequences can be finite or infinite. Two examples are the finite sequence (,2,0,) and the infinite sequence of odd numbers (1,3,5,7,9,). We use the notation an to represent the n-th term of a sequence.\n",
    "\n",
    "A recurrence relation is a way of defining the terms of a sequence with respect to the values of previous terms. In the case of Fibonacci's rabbits from the introduction, any given month will contain the rabbits that were alive the previous month, plus any new offspring. A key observation is that the number of offspring in any month is equal to the number of rabbits that were alive two months prior. As a result, if Fn represents the number of rabbit pairs alive after the n-th month, then we obtain the Fibonacci sequence having terms Fn that are defined by the recurrence relation Fn=Fn1+Fn2 (with F1=F2=1 to initiate the sequence). Although the sequence bears Fibonacci's name, it was known to Indian mathematicians over two millennia ago.\n",
    "\n",
    "When finding the n-th term of a sequence defined by a recurrence relation, we can simply use the recurrence relation to generate terms for progressively larger values of n. This problem introduces us to the computational technique of dynamic programming, which successively builds up solutions by using the answers to smaller cases.\n",
    "\n",
    "Given: Positive integers n40 and k5. E.g., 5 3\n",
    "\n",
    "Return: The total number of rabbit pairs that will be present after n months, if we begin with 1 pair and in each generation, every pair of reproduction-age rabbits produces a litter of k rabbit pairs (instead of only 1 pair). E.g., 19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "12bd9644-1416-46d6-8cb0-1c25e3c4ed24",
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 5\n",
    "k = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "b2e22386-7223-4fb7-80dd-e3f27834fa36",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fib(n, k=1): # if no value to k is given, k is 1\n",
    "    \n",
    "    if n == 1:\n",
    "        return 1\n",
    "    if n == 2:\n",
    "        return 1\n",
    "    # if n == 3:\n",
    "    #     return 1 + k\n",
    "        \n",
    "    return fib(n-1,k) + fib(n-2,k) * k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "4cba0733-8bd2-4afd-8958-a0a7e289d776",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fib(n,k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "406b292c-8299-42b8-bce4-87c87b0358a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "3\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "dictionary = {\n",
    "    0 : 1,\n",
    "    1 : 1,\n",
    "}\n",
    "n = 5\n",
    "k = 3\n",
    "\n",
    "for i in range(2, n, ):\n",
    "    dictionary[i] = \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e51c935-0b93-4f38-9b2c-5240eadca3d4",
   "metadata": {},
   "source": [
    "## Variables\n",
    "- n = input number of months\n",
    "- k = input number of offspring pairs to be produced by a mature pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "id": "74ec788e-8532-4f66-a0f7-5f8493938197",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "insert number of months: 5\n",
      "insert number of offsprings to be produced per pair: 3\n"
     ]
    }
   ],
   "source": [
    "n = int(input('insert number of months:'))\n",
    "k = int(input('insert number of offsprings to be produced per pair:'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b91e771-3065-48c7-ae1b-b613a7babc8e",
   "metadata": {},
   "source": [
    "## Programme\n",
    "\n",
    "    define function and its input variables, do:\n",
    "        add dictionary accumulator variable\n",
    "            for each step in range of n, do:\n",
    "                if step is equal to 0\n",
    "                    insert a \"step:1\" key pair into the dictionary; step = 0\n",
    "                else if step is equal to 1, do:\n",
    "                    insert a \"step:1\" key pair into the dictionary; step = 1\n",
    "                else, do:\n",
    "                    sum of pairs for the current step will be the value from the previous step in the dictionary and the one before previous step multiplied by number of pairs produced\n",
    "                   add the key pair consisting of the step and the sum of pairs value into the dictionary\n",
    "                return the number of pairs in the final month"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "id": "76c35ddf-9a1c-4daf-97fd-2d5fc308018e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fib_plus (n,k):\n",
    "    counter = {}\n",
    "    for step in range(n):\n",
    "        if step == 0:\n",
    "            counter.update({step:1})\n",
    "        elif step == 1:\n",
    "            counter.update({step:1})\n",
    "        else:\n",
    "            sum_pairs = counter[step-1] + counter[step-2] * k\n",
    "            counter.update({step:sum_pairs})\n",
    "    \n",
    "    return max(counter.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "8ba81089-347e-44c7-bd65-cbe7d4894daa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 490,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fib_plus(n,k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bef8c8a4-a238-4dd5-b5d2-d1107d1c82cb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
