{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68f9b39a",
   "metadata": {},
   "source": [
    "## ORF - Open Reading Frames (Loop/Con)\n",
    "### Theory\n",
    "In “Transcribing DNA into RNA”, we discussed the transcription of DNA into RNA, and in “Translating RNA into Protein”, we examined the translation of RNA into a chain of amino acids for the construction of proteins. We can view these two processes as a single step in which we directly translate a DNA string into a protein string, thus calling for a DNA codon table.\n",
    "\n",
    "However, three immediate wrinkles of complexity arise when we try to pass directly from DNA to proteins. First, not all DNA will be transcribed into RNA: so-called junk DNA appears to have no practical purpose for cellular function. Second, we can begin translation at any position along a strand of RNA, meaning that any substring of a DNA string can serve as a template for translation, as long as it begins with a start codon, ends with a stop codon, and has no other stop codons in the middle. As a result, the same RNA string can actually be translated in three different ways, depending on how we group triplets of symbols into codons. For example, ...AUGCUGAC... can be translated as ...AUGCUG..., ...UGCUGA..., and ...GCUGAC..., which will typically produce wildly different protein strings.\n",
    "\n",
    "### Exercise\n",
    "Either strand of a DNA double helix can serve as the coding strand for RNA transcription. Hence, a given DNA string implies six total reading frames, or ways in which the same region of DNA can be translated into amino acids: three reading frames result from reading the string itself, whereas three more result from reading its reverse complement.\n",
    "\n",
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "\n",
    "Given: A DNA string s of length at most 1 kbp in FASTA format.\n",
    "\n",
    "E.g.\n",
    "```\n",
    "\\>Rosalind_99\n",
    "AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG\n",
    "```\n",
    "Return: Every distinct candidate protein string that can be translated from ORFs of s. Strings can be returned in any order.\n",
    "\n",
    "E.g.\n",
    "```\n",
    "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
    "M\n",
    "MGMTPRLGLESLLE\n",
    "MTPRLGLESLLE\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4679042c",
   "metadata": {},
   "source": [
    "## Notes\n",
    "DNA double helix\n",
    "- sequence\n",
    "- reverse complement (opposite base pairs and reversed)\n",
    "    - each has 3 reading frames (we do not know where it starts; in total 6), starting from the 1st, 2nd and 3rd nucleotide base from the left\n",
    "        - keep unique translated sequences (we want all possibilities that start with start codon and end with an end codon in a given sequence, accounting for (relatively) unknown starting position)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d05e8f",
   "metadata": {},
   "source": [
    "## Input Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "66e199af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AGACTAACTGAATAATTCCCTCACCATGCAGCGGCCTAGACGACCTTGTTCGGATTCGGACTACGCTCACTCATACGAGTCACCGTTTGTCCGGGGGGTTATGTAACTTCGCAGGCGCGTAAGACGTCTTCACGCTTGCAGTTCGAGGGTCGGATCCGTGGAAGGCCAAGATGACTACCATGCATTATGCGTCGTGCCATGGCGTCATAATACCCGGAGCTAACAGTGACTTCAGCTTAGCGGTGCAATAGATCCATGTCAATATCCAGATGATCCGCCTCCTAGCGGCTATGAACAGATACCGGCCGAAATCAGGATTTACCAAAATGTTATCATCTAGTGTCAGGCAAGCCGTTCGGACAAAGAGCAGCGATGCAACCCAAAACCAGGAGAGCATCCGACGGCGCGGTAATGGTGGGCCAATTCACGGGTTTATGAGTTCTCGGGGTTAGTAGCTACTAACCCCGAGAACTCATACCCGATCCTAAAGATCCAACTGATTCCCGTGCCTGATGTGTAAAGGGTTCAAATGTTGCCTCCTCGGCCCAGTCGAAACAGGCGTCCCAACGTCGAGACATAATGTACCGTGCTCGGACTAGCCCCACCAAACCTAACCAATCTATTACCCGGGGTACGCTCCAAGGCTAACCCCGCATGTGGACCCAGATACCCGGAGCGGCGCCCAACCTGCCAAATCTTGTCATACCGTATCAAGTGCGTCCAACCTACTGGTCCTTAAGATCCGGGATCCGGCCACAGGTGAAAGGACTCGTTGTCAACATCTCGCGACGACCGCGATGTTCTCAACGGGATTGCTATAGAGTGCTTGCTTCGAACAACCGGCTCGGGGTTCGAAATTTCGGTTTTTACTTCACGCGTGGTGAGCTCATGATGGCGTAGTTTCGGCTTG']\n"
     ]
    }
   ],
   "source": [
    "# new version - used\n",
    "\n",
    "# insert a file path for DNA strings in FASTA format\n",
    "#path = str(insert('insert a file path:'))\n",
    "# fixed path\n",
    "path = '/mnt/c/Users/guspa/Desktop/python_homework/data/rosalind/rosalind_orf.txt'\n",
    "# reading mode\n",
    "mode = 'r'\n",
    "\n",
    "storage = []\n",
    "\n",
    "s = ''\n",
    "\n",
    "with open(path, mode) as file:\n",
    "    for line in file.readlines():\n",
    "        line = line.strip()\n",
    "\n",
    "        if '>' in line:\n",
    "            if s:\n",
    "                storage.append(s)\n",
    "                s = ''\n",
    "\n",
    "        else:\n",
    "            s +=  line\n",
    "\n",
    "if s:\n",
    "    storage.append(s)\n",
    "         \n",
    "print(storage) # lines with \">\" are not used, but are rather utilised to append whole (merged) sequences (not separated by new line characters)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "039d3e75",
   "metadata": {},
   "source": [
    "## Testing Ground"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "60070c2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG', 'CTGAGATGCTACTCGGATCATTCAGGCTTATTCCAAAAGAGACTCTAATCCAAGTCGCGGGGTCATCCCCATGTAACCTGAGTTAGCTACATGGCT']\n",
      "{'UUU': 'F', 'CUU': 'L', 'AUU': 'I', 'GUU': 'V', 'UUC': 'F', 'CUC': 'L', 'AUC': 'I', 'GUC': 'V', 'UUA': 'L', 'CUA': 'L', 'AUA': 'I', 'GUA': 'V', 'UUG': 'L', 'CUG': 'L', 'AUG': 'M', 'GUG': 'V', 'UCU': 'S', 'CCU': 'P', 'ACU': 'T', 'GCU': 'A', 'UCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A', 'UCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A', 'UCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A', 'UAU': 'Y', 'CAU': 'H', 'AAU': 'N', 'GAU': 'D', 'UAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D', 'UAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E', 'UAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E', 'UGU': 'C', 'CGU': 'R', 'AGU': 'S', 'GGU': 'G', 'UGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G', 'UGA': 'Stop', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G', 'UGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'}\n",
      "{'TTT': 'F', 'CTT': 'L', 'ATT': 'I', 'GTT': 'V', 'TTC': 'F', 'CTC': 'L', 'ATC': 'I', 'GTC': 'V', 'TTA': 'L', 'CTA': 'L', 'ATA': 'I', 'GTA': 'V', 'TTG': 'L', 'CTG': 'L', 'ATG': 'M', 'GTG': 'V', 'TCT': 'S', 'CCT': 'P', 'ACT': 'T', 'GCT': 'A', 'TCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A', 'TCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A', 'TCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A', 'TAT': 'Y', 'CAT': 'H', 'AAT': 'N', 'GAT': 'D', 'TAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D', 'TAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E', 'TAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E', 'TGT': 'C', 'CGT': 'R', 'AGT': 'S', 'GGT': 'G', 'TGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G', 'TGA': 'Stop', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G', 'TGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'}\n"
     ]
    }
   ],
   "source": [
    "storage_test = ['AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG', 'CTGAGATGCTACTCGGATCATTCAGGCTTATTCCAAAAGAGACTCTAATCCAAGTCGCGGGGTCATCCCCATGTAACCTGAGTTAGCTACATGGCT']\n",
    "\n",
    "print(storage_test)\n",
    "\n",
    "# not used\n",
    "combinations_rna = {\n",
    "    'UUU' : 'F',    'CUU' : 'L', 'AUU' : 'I', 'GUU' : 'V',\n",
    "    'UUC' : 'F',    'CUC' : 'L', 'AUC' : 'I', 'GUC' : 'V',\n",
    "    'UUA' : 'L',    'CUA' : 'L', 'AUA' : 'I', 'GUA' : 'V',\n",
    "    'UUG' : 'L',    'CUG' : 'L', 'AUG' : 'M', 'GUG' : 'V',\n",
    "    'UCU' : 'S',    'CCU' : 'P', 'ACU' : 'T', 'GCU' : 'A',\n",
    "    'UCC' : 'S',    'CCC' : 'P', 'ACC' : 'T', 'GCC' : 'A',\n",
    "    'UCA' : 'S',    'CCA' : 'P', 'ACA' : 'T', 'GCA' : 'A',\n",
    "    'UCG' : 'S',    'CCG' : 'P', 'ACG' : 'T', 'GCG' : 'A',\n",
    "    'UAU' : 'Y',    'CAU' : 'H', 'AAU' : 'N', 'GAU' : 'D',\n",
    "    'UAC' : 'Y',    'CAC' : 'H', 'AAC' : 'N', 'GAC' : 'D',\n",
    "    'UAA' : 'Stop', 'CAA' : 'Q', 'AAA' : 'K', 'GAA' : 'E',\n",
    "    'UAG' : 'Stop', 'CAG' : 'Q', 'AAG' : 'K', 'GAG' : 'E',\n",
    "    'UGU' : 'C',    'CGU' : 'R', 'AGU' : 'S', 'GGU' : 'G',\n",
    "    'UGC' : 'C',    'CGC' : 'R', 'AGC' : 'S', 'GGC' : 'G',\n",
    "    'UGA' : 'Stop', 'CGA' : 'R', 'AGA' : 'R', 'GGA' : 'G',\n",
    "    'UGG' : 'W',    'CGG' : 'R', 'AGG' : 'R', 'GGG' : 'G'\n",
    "}\n",
    "\n",
    "# used\n",
    "combinations_dna = {\n",
    "    'TTT' : 'F',      'CTT' : 'L',      'ATT' : 'I',      'GTT' : 'V',\n",
    "    'TTC' : 'F',      'CTC' : 'L',      'ATC' : 'I',      'GTC' : 'V',\n",
    "    'TTA' : 'L',      'CTA' : 'L',      'ATA' : 'I',      'GTA' : 'V',\n",
    "    'TTG' : 'L',      'CTG' : 'L',      'ATG' : 'M',      'GTG' : 'V',\n",
    "    'TCT' : 'S',      'CCT' : 'P',      'ACT' : 'T',      'GCT' : 'A',\n",
    "    'TCC' : 'S',      'CCC' : 'P',      'ACC' : 'T',      'GCC' : 'A',\n",
    "    'TCA' : 'S',      'CCA' : 'P',      'ACA' : 'T',      'GCA' : 'A',\n",
    "    'TCG' : 'S',      'CCG' : 'P',      'ACG' : 'T',      'GCG' : 'A',\n",
    "    'TAT' : 'Y',      'CAT' : 'H',      'AAT' : 'N',      'GAT' : 'D',\n",
    "    'TAC' : 'Y',      'CAC' : 'H',      'AAC' : 'N',      'GAC' : 'D',\n",
    "    'TAA' : 'Stop',   'CAA' : 'Q',      'AAA' : 'K',      'GAA' : 'E',\n",
    "    'TAG' : 'Stop',   'CAG' : 'Q',      'AAG' : 'K',      'GAG' : 'E',\n",
    "    'TGT' : 'C',      'CGT' : 'R',      'AGT' : 'S',      'GGT' : 'G',\n",
    "    'TGC' : 'C',      'CGC' : 'R',      'AGC' : 'S',      'GGC' : 'G',\n",
    "    'TGA' : 'Stop',   'CGA' : 'R',      'AGA' : 'R',      'GGA' : 'G',\n",
    "    'TGG' : 'W',      'CGG' : 'R',      'AGG' : 'R',      'GGG' : 'G' \n",
    "}\n",
    "\n",
    "print(combinations_rna); print(combinations_dna)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02214ec9",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d5ecc63f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCCAUGUAGCUAACUCAGGUUACAUGGGGAUGACCCCGCGACUUGGAUUAGAGUCUCUUUUGGAAUAAGCCUGAAUGAUCCGAGUAGCAUCUCAG\n"
     ]
    }
   ],
   "source": [
    "# not used \n",
    "def transcription(string):\n",
    "\n",
    "    transcribed_string = ''\n",
    "\n",
    "    for nucleotide_base in string:\n",
    "\n",
    "        if nucleotide_base == 'T':\n",
    "            transcribed_string += 'U'\n",
    "        \n",
    "        else:\n",
    "            transcribed_string += nucleotide_base\n",
    "\n",
    "    return transcribed_string\n",
    "\n",
    "rna_sequence = transcription(storage_test[0]); print(rna_sequence) # changes thymine (dna) to uracil (rna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "483a85f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTGAGATGCTACTCGGATCATTCAGGCTTATTCCAAAAGAGACTCTAATCCAAGTCGCGGGGTCATCCCCATGTAACCTGAGTTAGCTACATGGCT\n"
     ]
    }
   ],
   "source": [
    "def reverse(string):\n",
    "\n",
    "    # create a sequence with complementing base pairs\n",
    "    reversed_string = string.replace('A', 't')\n",
    "    reversed_string = reversed_string.replace('T', 'a')\n",
    "    reversed_string = reversed_string.replace('C', 'g')\n",
    "    reversed_string = reversed_string.replace('G', 'c')\n",
    "\n",
    "    reversed_string = reversed_string.upper() # capitalise letters\n",
    "\n",
    "    reversed_string = reversed_string[::-1] # reverse order\n",
    "\n",
    "    return reversed_string\n",
    "\n",
    "reverse_complement = reverse(storage_test[0]); print(reverse_complement)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "81ff0a83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['AGC', 'CAT', 'GTA', 'GCT', 'AAC', 'TCA', 'GGT', 'TAC', 'ATG', 'GGG', 'ATG', 'ACC', 'CCG', 'CGA', 'CTT', 'GGA', 'TTA', 'GAG', 'TCT', 'CTT', 'TTG', 'GAA', 'TAA', 'GCC', 'TGA', 'ATG', 'ATC', 'CGA', 'GTA', 'GCA', 'TCT', 'CAG'], ['GCC', 'ATG', 'TAG', 'CTA', 'ACT', 'CAG', 'GTT', 'ACA', 'TGG', 'GGA', 'TGA', 'CCC', 'CGC', 'GAC', 'TTG', 'GAT', 'TAG', 'AGT', 'CTC', 'TTT', 'TGG', 'AAT', 'AAG', 'CCT', 'GAA', 'TGA', 'TCC', 'GAG', 'TAG', 'CAT', 'CTC'], ['CCA', 'TGT', 'AGC', 'TAA', 'CTC', 'AGG', 'TTA', 'CAT', 'GGG', 'GAT', 'GAC', 'CCC', 'GCG', 'ACT', 'TGG', 'ATT', 'AGA', 'GTC', 'TCT', 'TTT', 'GGA', 'ATA', 'AGC', 'CTG', 'AAT', 'GAT', 'CCG', 'AGT', 'AGC', 'ATC', 'TCA']]\n",
      "\n",
      "[['AGA', 'CTA', 'ACT', 'GAA', 'TAA', 'TTC', 'CCT', 'CAC', 'CAT', 'GCA', 'GCG', 'GCC', 'TAG', 'ACG', 'ACC', 'TTG', 'TTC', 'GGA', 'TTC', 'GGA', 'CTA', 'CGC', 'TCA', 'CTC', 'ATA', 'CGA', 'GTC', 'ACC', 'GTT', 'TGT', 'CCG', 'GGG', 'GGT', 'TAT', 'GTA', 'ACT', 'TCG', 'CAG', 'GCG', 'CGT', 'AAG', 'ACG', 'TCT', 'TCA', 'CGC', 'TTG', 'CAG', 'TTC', 'GAG', 'GGT', 'CGG', 'ATC', 'CGT', 'GGA', 'AGG', 'CCA', 'AGA', 'TGA', 'CTA', 'CCA', 'TGC', 'ATT', 'ATG', 'CGT', 'CGT', 'GCC', 'ATG', 'GCG', 'TCA', 'TAA', 'TAC', 'CCG', 'GAG', 'CTA', 'ACA', 'GTG', 'ACT', 'TCA', 'GCT', 'TAG', 'CGG', 'TGC', 'AAT', 'AGA', 'TCC', 'ATG', 'TCA', 'ATA', 'TCC', 'AGA', 'TGA', 'TCC', 'GCC', 'TCC', 'TAG', 'CGG', 'CTA', 'TGA', 'ACA', 'GAT', 'ACC', 'GGC', 'CGA', 'AAT', 'CAG', 'GAT', 'TTA', 'CCA', 'AAA', 'TGT', 'TAT', 'CAT', 'CTA', 'GTG', 'TCA', 'GGC', 'AAG', 'CCG', 'TTC', 'GGA', 'CAA', 'AGA', 'GCA', 'GCG', 'ATG', 'CAA', 'CCC', 'AAA', 'ACC', 'AGG', 'AGA', 'GCA', 'TCC', 'GAC', 'GGC', 'GCG', 'GTA', 'ATG', 'GTG', 'GGC', 'CAA', 'TTC', 'ACG', 'GGT', 'TTA', 'TGA', 'GTT', 'CTC', 'GGG', 'GTT', 'AGT', 'AGC', 'TAC', 'TAA', 'CCC', 'CGA', 'GAA', 'CTC', 'ATA', 'CCC', 'GAT', 'CCT', 'AAA', 'GAT', 'CCA', 'ACT', 'GAT', 'TCC', 'CGT', 'GCC', 'TGA', 'TGT', 'GTA', 'AAG', 'GGT', 'TCA', 'AAT', 'GTT', 'GCC', 'TCC', 'TCG', 'GCC', 'CAG', 'TCG', 'AAA', 'CAG', 'GCG', 'TCC', 'CAA', 'CGT', 'CGA', 'GAC', 'ATA', 'ATG', 'TAC', 'CGT', 'GCT', 'CGG', 'ACT', 'AGC', 'CCC', 'ACC', 'AAA', 'CCT', 'AAC', 'CAA', 'TCT', 'ATT', 'ACC', 'CGG', 'GGT', 'ACG', 'CTC', 'CAA', 'GGC', 'TAA', 'CCC', 'CGC', 'ATG', 'TGG', 'ACC', 'CAG', 'ATA', 'CCC', 'GGA', 'GCG', 'GCG', 'CCC', 'AAC', 'CTG', 'CCA', 'AAT', 'CTT', 'GTC', 'ATA', 'CCG', 'TAT', 'CAA', 'GTG', 'CGT', 'CCA', 'ACC', 'TAC', 'TGG', 'TCC', 'TTA', 'AGA', 'TCC', 'GGG', 'ATC', 'CGG', 'CCA', 'CAG', 'GTG', 'AAA', 'GGA', 'CTC', 'GTT', 'GTC', 'AAC', 'ATC', 'TCG', 'CGA', 'CGA', 'CCG', 'CGA', 'TGT', 'TCT', 'CAA', 'CGG', 'GAT', 'TGC', 'TAT', 'AGA', 'GTG', 'CTT', 'GCT', 'TCG', 'AAC', 'AAC', 'CGG', 'CTC', 'GGG', 'GTT', 'CGA', 'AAT', 'TTC', 'GGT', 'TTT', 'TAC', 'TTC', 'ACG', 'CGT', 'GGT', 'GAG', 'CTC', 'ATG', 'ATG', 'GCG', 'TAG', 'TTT', 'CGG', 'CTT'], ['GAC', 'TAA', 'CTG', 'AAT', 'AAT', 'TCC', 'CTC', 'ACC', 'ATG', 'CAG', 'CGG', 'CCT', 'AGA', 'CGA', 'CCT', 'TGT', 'TCG', 'GAT', 'TCG', 'GAC', 'TAC', 'GCT', 'CAC', 'TCA', 'TAC', 'GAG', 'TCA', 'CCG', 'TTT', 'GTC', 'CGG', 'GGG', 'GTT', 'ATG', 'TAA', 'CTT', 'CGC', 'AGG', 'CGC', 'GTA', 'AGA', 'CGT', 'CTT', 'CAC', 'GCT', 'TGC', 'AGT', 'TCG', 'AGG', 'GTC', 'GGA', 'TCC', 'GTG', 'GAA', 'GGC', 'CAA', 'GAT', 'GAC', 'TAC', 'CAT', 'GCA', 'TTA', 'TGC', 'GTC', 'GTG', 'CCA', 'TGG', 'CGT', 'CAT', 'AAT', 'ACC', 'CGG', 'AGC', 'TAA', 'CAG', 'TGA', 'CTT', 'CAG', 'CTT', 'AGC', 'GGT', 'GCA', 'ATA', 'GAT', 'CCA', 'TGT', 'CAA', 'TAT', 'CCA', 'GAT', 'GAT', 'CCG', 'CCT', 'CCT', 'AGC', 'GGC', 'TAT', 'GAA', 'CAG', 'ATA', 'CCG', 'GCC', 'GAA', 'ATC', 'AGG', 'ATT', 'TAC', 'CAA', 'AAT', 'GTT', 'ATC', 'ATC', 'TAG', 'TGT', 'CAG', 'GCA', 'AGC', 'CGT', 'TCG', 'GAC', 'AAA', 'GAG', 'CAG', 'CGA', 'TGC', 'AAC', 'CCA', 'AAA', 'CCA', 'GGA', 'GAG', 'CAT', 'CCG', 'ACG', 'GCG', 'CGG', 'TAA', 'TGG', 'TGG', 'GCC', 'AAT', 'TCA', 'CGG', 'GTT', 'TAT', 'GAG', 'TTC', 'TCG', 'GGG', 'TTA', 'GTA', 'GCT', 'ACT', 'AAC', 'CCC', 'GAG', 'AAC', 'TCA', 'TAC', 'CCG', 'ATC', 'CTA', 'AAG', 'ATC', 'CAA', 'CTG', 'ATT', 'CCC', 'GTG', 'CCT', 'GAT', 'GTG', 'TAA', 'AGG', 'GTT', 'CAA', 'ATG', 'TTG', 'CCT', 'CCT', 'CGG', 'CCC', 'AGT', 'CGA', 'AAC', 'AGG', 'CGT', 'CCC', 'AAC', 'GTC', 'GAG', 'ACA', 'TAA', 'TGT', 'ACC', 'GTG', 'CTC', 'GGA', 'CTA', 'GCC', 'CCA', 'CCA', 'AAC', 'CTA', 'ACC', 'AAT', 'CTA', 'TTA', 'CCC', 'GGG', 'GTA', 'CGC', 'TCC', 'AAG', 'GCT', 'AAC', 'CCC', 'GCA', 'TGT', 'GGA', 'CCC', 'AGA', 'TAC', 'CCG', 'GAG', 'CGG', 'CGC', 'CCA', 'ACC', 'TGC', 'CAA', 'ATC', 'TTG', 'TCA', 'TAC', 'CGT', 'ATC', 'AAG', 'TGC', 'GTC', 'CAA', 'CCT', 'ACT', 'GGT', 'CCT', 'TAA', 'GAT', 'CCG', 'GGA', 'TCC', 'GGC', 'CAC', 'AGG', 'TGA', 'AAG', 'GAC', 'TCG', 'TTG', 'TCA', 'ACA', 'TCT', 'CGC', 'GAC', 'GAC', 'CGC', 'GAT', 'GTT', 'CTC', 'AAC', 'GGG', 'ATT', 'GCT', 'ATA', 'GAG', 'TGC', 'TTG', 'CTT', 'CGA', 'ACA', 'ACC', 'GGC', 'TCG', 'GGG', 'TTC', 'GAA', 'ATT', 'TCG', 'GTT', 'TTT', 'ACT', 'TCA', 'CGC', 'GTG', 'GTG', 'AGC', 'TCA', 'TGA', 'TGG', 'CGT', 'AGT', 'TTC', 'GGC', 'TTG'], ['ACT', 'AAC', 'TGA', 'ATA', 'ATT', 'CCC', 'TCA', 'CCA', 'TGC', 'AGC', 'GGC', 'CTA', 'GAC', 'GAC', 'CTT', 'GTT', 'CGG', 'ATT', 'CGG', 'ACT', 'ACG', 'CTC', 'ACT', 'CAT', 'ACG', 'AGT', 'CAC', 'CGT', 'TTG', 'TCC', 'GGG', 'GGG', 'TTA', 'TGT', 'AAC', 'TTC', 'GCA', 'GGC', 'GCG', 'TAA', 'GAC', 'GTC', 'TTC', 'ACG', 'CTT', 'GCA', 'GTT', 'CGA', 'GGG', 'TCG', 'GAT', 'CCG', 'TGG', 'AAG', 'GCC', 'AAG', 'ATG', 'ACT', 'ACC', 'ATG', 'CAT', 'TAT', 'GCG', 'TCG', 'TGC', 'CAT', 'GGC', 'GTC', 'ATA', 'ATA', 'CCC', 'GGA', 'GCT', 'AAC', 'AGT', 'GAC', 'TTC', 'AGC', 'TTA', 'GCG', 'GTG', 'CAA', 'TAG', 'ATC', 'CAT', 'GTC', 'AAT', 'ATC', 'CAG', 'ATG', 'ATC', 'CGC', 'CTC', 'CTA', 'GCG', 'GCT', 'ATG', 'AAC', 'AGA', 'TAC', 'CGG', 'CCG', 'AAA', 'TCA', 'GGA', 'TTT', 'ACC', 'AAA', 'ATG', 'TTA', 'TCA', 'TCT', 'AGT', 'GTC', 'AGG', 'CAA', 'GCC', 'GTT', 'CGG', 'ACA', 'AAG', 'AGC', 'AGC', 'GAT', 'GCA', 'ACC', 'CAA', 'AAC', 'CAG', 'GAG', 'AGC', 'ATC', 'CGA', 'CGG', 'CGC', 'GGT', 'AAT', 'GGT', 'GGG', 'CCA', 'ATT', 'CAC', 'GGG', 'TTT', 'ATG', 'AGT', 'TCT', 'CGG', 'GGT', 'TAG', 'TAG', 'CTA', 'CTA', 'ACC', 'CCG', 'AGA', 'ACT', 'CAT', 'ACC', 'CGA', 'TCC', 'TAA', 'AGA', 'TCC', 'AAC', 'TGA', 'TTC', 'CCG', 'TGC', 'CTG', 'ATG', 'TGT', 'AAA', 'GGG', 'TTC', 'AAA', 'TGT', 'TGC', 'CTC', 'CTC', 'GGC', 'CCA', 'GTC', 'GAA', 'ACA', 'GGC', 'GTC', 'CCA', 'ACG', 'TCG', 'AGA', 'CAT', 'AAT', 'GTA', 'CCG', 'TGC', 'TCG', 'GAC', 'TAG', 'CCC', 'CAC', 'CAA', 'ACC', 'TAA', 'CCA', 'ATC', 'TAT', 'TAC', 'CCG', 'GGG', 'TAC', 'GCT', 'CCA', 'AGG', 'CTA', 'ACC', 'CCG', 'CAT', 'GTG', 'GAC', 'CCA', 'GAT', 'ACC', 'CGG', 'AGC', 'GGC', 'GCC', 'CAA', 'CCT', 'GCC', 'AAA', 'TCT', 'TGT', 'CAT', 'ACC', 'GTA', 'TCA', 'AGT', 'GCG', 'TCC', 'AAC', 'CTA', 'CTG', 'GTC', 'CTT', 'AAG', 'ATC', 'CGG', 'GAT', 'CCG', 'GCC', 'ACA', 'GGT', 'GAA', 'AGG', 'ACT', 'CGT', 'TGT', 'CAA', 'CAT', 'CTC', 'GCG', 'ACG', 'ACC', 'GCG', 'ATG', 'TTC', 'TCA', 'ACG', 'GGA', 'TTG', 'CTA', 'TAG', 'AGT', 'GCT', 'TGC', 'TTC', 'GAA', 'CAA', 'CCG', 'GCT', 'CGG', 'GGT', 'TCG', 'AAA', 'TTT', 'CGG', 'TTT', 'TTA', 'CTT', 'CAC', 'GCG', 'TGG', 'TGA', 'GCT', 'CAT', 'GAT', 'GGC', 'GTA', 'GTT', 'TCG', 'GCT']]\n"
     ]
    }
   ],
   "source": [
    "def get_codons(string):\n",
    "\n",
    "    codons = [[], [], []]\n",
    "\n",
    "    if len(string) % 3 == 1:\n",
    "        # for each open reading frame (3 in total)\n",
    "        for orf in range(3): # 0, 1, 2 (offset)\n",
    "\n",
    "            # correction: -1, 0, -2 (removal of nucleotide bases from the end - junk DNA)\n",
    "            if orf == 0:\n",
    "                correction = -1  \n",
    "            elif orf == 1:\n",
    "                correction = 0\n",
    "            else:\n",
    "                correction = -2\n",
    "            \n",
    "        # acquire triplets \n",
    "            for codon in range((len(string) + correction) // 3):\n",
    "\n",
    "                next_codon = codon * 3\n",
    "\n",
    "                start = next_codon + orf\n",
    "                end = next_codon + orf + 3\n",
    "\n",
    "                codons[orf].append(string[start:end])\n",
    "\n",
    "    elif len(string) % 3 == 2:\n",
    "        # for each open reading frame (3 in total)\n",
    "        for orf in range(3): # 0, 1, 2 (offset)\n",
    "\n",
    "            # correction: -2, -1, 0 (removal of nucleotide bases from the end - junk DNA)\n",
    "            if orf == 0:\n",
    "                correction = -2  \n",
    "            elif orf == 1:\n",
    "                correction = -1\n",
    "            else:\n",
    "                correction = 0\n",
    "            \n",
    "        # acquire triplets \n",
    "            for codon in range((len(string) + correction) // 3):\n",
    "\n",
    "                next_codon = codon * 3\n",
    "\n",
    "                start = next_codon + orf\n",
    "                end = next_codon + orf + 3\n",
    "\n",
    "                codons[orf].append(string[start:end])\n",
    "\n",
    "    else: \n",
    "        # for each open reading frame (3 in total)\n",
    "        for orf in range(3): # 0, 1, 2 (offset)\n",
    "\n",
    "            # correction: 0, -2, -1 (removal of nucleotide bases from the end - junk DNA)\n",
    "            if orf == 0:\n",
    "                correction = 0  \n",
    "            elif orf == 1:\n",
    "                correction = -2\n",
    "            else:\n",
    "                correction = -1\n",
    "\n",
    "            # acquire triplets \n",
    "            for codon in range((len(string) + correction) // 3):\n",
    "\n",
    "                next_codon = codon * 3\n",
    "\n",
    "                start = next_codon + orf\n",
    "                end = next_codon + orf + 3\n",
    "\n",
    "                codons[orf].append(string[start:end])\n",
    "\n",
    "    return codons\n",
    "\n",
    "print(get_codons(storage_test[0]))\n",
    "print()\n",
    "print(get_codons(storage[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "253ba497",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MGMTPRLGLESLLE', 'MTPRLGLESLLE', 'M']\n",
      "\n",
      "['M', 'MLLGSFRLIPKETLIQVAGSSPCNLS']\n"
     ]
    }
   ],
   "source": [
    "def translation(codon_list, codon_combinations):\n",
    "\n",
    "    proteins = []\n",
    "    protein_sequence = ''\n",
    "\n",
    "    # loop through codons\n",
    "    for orf in range(len(codon_list)):\n",
    "        for position_dna, codon in enumerate(codon_list[orf]):\n",
    "\n",
    "            # locate start codon\n",
    "            if codon == 'ATG':\n",
    "\n",
    "                # loop through next codons, beginning with the start codon and translate them until the stop codon is reached\n",
    "                for codon_in_sequence in codon_list[orf][position_dna : ]:\n",
    "\n",
    "                    translation = codon_combinations[codon_in_sequence]\n",
    "\n",
    "                    if translation != 'Stop':\n",
    "                        protein_sequence += translation\n",
    "                \n",
    "                    else:\n",
    "                        proteins.append(protein_sequence)\n",
    "                        protein_sequence = ''\n",
    "\n",
    "                        break\n",
    "\n",
    "            protein_sequence = ''\n",
    "\n",
    "    return proteins\n",
    "\n",
    "print(translation(get_codons(storage_test[0]), combinations_dna))\n",
    "print()\n",
    "print(translation(get_codons(storage_test[1]), combinations_dna))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17892468",
   "metadata": {},
   "source": [
    "## Programme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ec41a786",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRRAMAS\n",
      "MAS\n",
      "MSISR\n",
      "MQPKTRRASDGAVMVGQFTGL\n",
      "MVGQFTGL\n",
      "MYRARTSPTKPNQSITRGTLQG\n",
      "MWTQIPGAAPNLPNLVIPYQVRPTYWSLRSGIRPQVKGLVVNISRRPRCSQRDCYRVLASNNRLGVRNFGFYFTRGELMMA\n",
      "MMA\n",
      "MA\n",
      "MQRPRRPCSDSDYAHSYESPFVRGVM\n",
      "M\n",
      "MLPPRPSRNRRPNVET\n",
      "MTTMHYASCHGVIIPGANSDFSLAVQ\n",
      "MHYASCHGVIIPGANSDFSLAVQ\n",
      "MIRLLAAMNRYRPKSGFTKMLSSSVRQAVRTKSSDATQNQESIRRRGNGGPIHGFMSSRG\n",
      "MNRYRPKSGFTKMLSSSVRQAVRTKSSDATQNQESIRRRGNGGPIHGFMSSRG\n",
      "MLSSSVRQAVRTKSSDATQNQESIRRRGNGGPIHGFMSSRG\n",
      "MSSRG\n",
      "MCKGFKCCLLGPVETGVPTSRHNVPCSD\n",
      "MFSTGLL\n",
      "MLSWFWVASLLFVRTACLTLDDNILVNPDFGRYLFIAARRRIIWILTWIYCTAKLKSLLAPGIMTPWHDA\n",
      "MTPWHDA\n",
      "MVVILAFHGSDPRTASVKTSYAPAKLHNPPDKR\n",
      "MITFW\n",
      "MVRELFS\n",
      "MSSPRVK\n",
      "MLTTSPFTCGRIPDLKDQ\n",
      "MTRFGRLGAAPGIWVHMRG\n",
      "MRG\n",
      "MSRRWDACFDWAEEATFEPFTHQARESVGSLGSGMSSRG\n",
      "MDLLHR\n",
      "MARRIMHGSHLGLPRIRPSNCKREDVLRACEVT\n",
      "MHGSHLGLPRIRPSNCKREDVLRACEVT\n",
      "MSERSPNPNKVV\n"
     ]
    }
   ],
   "source": [
    "# get all possible protein sequences from the open reading frames\n",
    "proteins = []\n",
    "\n",
    "for dna_sequence in storage: # storage_test or storage\n",
    "    \n",
    "    reverse_complement = reverse(dna_sequence)\n",
    "\n",
    "    dna_codons = get_codons(dna_sequence)\n",
    "    dna_rc_codons = get_codons(reverse_complement)\n",
    "\n",
    "    proteins.append(translation(dna_codons, combinations_dna))\n",
    "    proteins.append(translation(dna_rc_codons, combinations_dna))\n",
    "    \n",
    "\n",
    "# keep unique protein sequences only\n",
    "unique_proteins = []\n",
    "\n",
    "for iteration in range(len(proteins)):\n",
    "    for protein in proteins[iteration]:\n",
    "\n",
    "        if protein not in unique_proteins:\n",
    "            unique_proteins.append(protein)\n",
    "\n",
    "\n",
    "# print the results\n",
    "for protein in unique_proteins:\n",
    "    print(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "464818b9",
   "metadata": {},
   "source": [
    "## Answer\n",
    "- can be in any order\n",
    "\n",
    "```\n",
    "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
    "M\n",
    "MGMTPRLGLESLLE\n",
    "MTPRLGLESLLE\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
